// @alifd/meet-react/lib/message/func.js
"use strict";

exports.__esModule = true;
exports.success = exports.show = exports.notice = exports.loading = exports.hide = exports.help = exports.error = void 0;
var _tslib = require("tslib");
var _react = _interopRequireWildcard(require("react"));
var _render = require("../utils/render");
var _message = _interopRequireDefault(require("./message"));
var _view = _interopRequireDefault(require("../view"));
var _utils = require("../utils");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
var hideCurrent = null;
var hide = function () {
  if (hideCurrent) {
    hideCurrent();
  }
  hideCurrent = null;
};
exports.hide = hide;
var show = function (options) {
  var _options$shape = options.shape,
    shape = _options$shape === void 0 ? 'toast' : _options$shape,
    _options$prefix = options.prefix,
    prefix = _options$prefix === void 0 ? 'mt-' : _options$prefix,
    _options$duration = options.duration,
    duration = _options$duration === void 0 ? 1500 : _options$duration,
    content = options.content,
    type = options.type,
    onClose = options.onClose,
    other = (0, _tslib.__rest)(options, ["shape", "prefix", "duration", "content", "type", "onClose"]);
  hide();
  var hideThis = (0, _render.renderToRoot)( /*#__PURE__*/(0, _react.createElement)(_view.default, {
    "disable-scroll": true
  }, /*#__PURE__*/(0, _react.createElement)(_message.default, _extends({
    visible: true,
    shape: shape,
    type: type
  }, other), content), /*#__PURE__*/(0, _react.createElement)(_view.default, {
    "disable-scroll": true,
    className: prefix + "message" + "-mask"
  })), onClose);
  hideCurrent = hideThis;
  if (duration > 0) {
    if (type === 'loading' && !('duration' in options)) {
      return hideThis;
    }
    setTimeout(function () {
      hideThis();
      if (hideCurrent === hideThis) {
        hideCurrent = null;
      }
    }, duration);
  }
  return hideThis;
};
exports.show = show;
var handleFastCall = function (type, options) {
  var opts;
  if ((0, _utils.isString)(options)) {
    opts = {
      title: options
    };
  } else {
    opts = options;
  }
  return show(Object.assign(Object.assign({}, opts), {
    type: type
  }));
};
var success = exports.success = handleFastCall.bind(null, 'success');
var error = exports.error = handleFastCall.bind(null, 'error');
var notice = exports.notice = handleFastCall.bind(null, 'notice');
var help = exports.help = handleFastCall.bind(null, 'help');
var loading = exports.loading = handleFastCall.bind(null, 'loading');
