// @alilc/lowcode-designer/lib/document/node/exclusive-group.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.ExclusiveGroup = void 0;
exports.isExclusiveGroup = isExclusiveGroup;
var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));
var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _lowcodeUtils = require("@alilc/lowcode-utils");
var _locale = require("../../locale");
var _dec, _class, _descriptor, _descriptor2;
// modals assoc x-hide value, initial: check is Modal, yes will put it in modals, cross levels
// if-else-if assoc conditionGroup value, should be the same level,
// and siblings, need renderEngine support
var ExclusiveGroup = exports.ExclusiveGroup = (_dec = _lowcodeEditorCore.obx.shallow, (_class = /*#__PURE__*/function () {
  function ExclusiveGroup(name, title) {
    this.name = name;
    this.isExclusiveGroup = true;
    this.id = (0, _lowcodeUtils.uniqueId)('exclusive');
    this.title = void 0;
    (0, _initializerDefineProperty2["default"])(this, "children", _descriptor, this);
    (0, _initializerDefineProperty2["default"])(this, "visibleIndex", _descriptor2, this);
    (0, _lowcodeEditorCore.makeObservable)(this);
    this.title = title || {
      type: 'i18n',
      intl: (0, _locale.intl)('Condition Group')
    };
  }
  var _proto = ExclusiveGroup.prototype;
  _proto.add = function add(node) {
    var _node$nextSibling$con;
    if (node.nextSibling && ((_node$nextSibling$con = node.nextSibling.conditionGroup) === null || _node$nextSibling$con === void 0 ? void 0 : _node$nextSibling$con.id) === this.id) {
      var i = this.children.indexOf(node.nextSibling);
      this.children.splice(i, 0, node);
    } else {
      this.children.push(node);
    }
  };
  _proto.remove = function remove(node) {
    var i = this.children.indexOf(node);
    if (i > -1) {
      this.children.splice(i, 1);
      if (this.visibleIndex > i) {
        this.visibleIndex -= 1;
      } else if (this.visibleIndex >= this.children.length) {
        this.visibleIndex = this.children.length - 1;
      }
    }
  };
  _proto.setVisible = function setVisible(node) {
    var i = this.children.indexOf(node);
    if (i > -1) {
      this.visibleIndex = i;
    }
  };
  _proto.isVisible = function isVisible(node) {
    var i = this.children.indexOf(node);
    return i === this.visibleIndex;
  };
  (0, _createClass2["default"])(ExclusiveGroup, [{
    key: "document",
    get: function get() {
      return this.visibleNode.document;
    }
  }, {
    key: "zLevel",
    get: function get() {
      return this.visibleNode.zLevel;
    }
  }, {
    key: "length",
    get: function get() {
      return this.children.length;
    }
  }, {
    key: "visibleNode",
    get: function get() {
      return this.children[this.visibleIndex];
    }
  }, {
    key: "firstNode",
    get: function get() {
      return this.children[0];
    }
  }, {
    key: "index",
    get: function get() {
      return this.firstNode.index;
    }
  }]);
  return ExclusiveGroup;
}(), (_descriptor = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "children", [_dec], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return [];
  }
}), _descriptor2 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "visibleIndex", [_lowcodeEditorCore.obx], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return 0;
  }
}), (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "document", [_lowcodeEditorCore.computed], Object.getOwnPropertyDescriptor(_class.prototype, "document"), _class.prototype), (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "zLevel", [_lowcodeEditorCore.computed], Object.getOwnPropertyDescriptor(_class.prototype, "zLevel"), _class.prototype), (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "length", [_lowcodeEditorCore.computed], Object.getOwnPropertyDescriptor(_class.prototype, "length"), _class.prototype), (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "visibleNode", [_lowcodeEditorCore.computed], Object.getOwnPropertyDescriptor(_class.prototype, "visibleNode"), _class.prototype), (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "firstNode", [_lowcodeEditorCore.computed], Object.getOwnPropertyDescriptor(_class.prototype, "firstNode"), _class.prototype)), _class));
function isExclusiveGroup(obj) {
  return obj && obj.isExclusiveGroup;
}
