// @alilc/lowcode-designer/lib/designer/active-tracker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.ActiveTracker = void 0;
var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));
var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _lowcodeUtils = require("@alilc/lowcode-utils");
var _dec, _class, _descriptor;
var ActiveTracker = exports.ActiveTracker = (_dec = _lowcodeEditorCore.obx.ref, (_class = /*#__PURE__*/function () {
  function ActiveTracker() {
    (0, _initializerDefineProperty2["default"])(this, "_target", _descriptor, this);
    this.emitter = (0, _lowcodeEditorCore.createModuleEventBus)('ActiveTracker');
  }
  var _proto = ActiveTracker.prototype;
  _proto.track = function track(originalTarget) {
    var target = originalTarget;
    if ((0, _lowcodeUtils.isNode)(originalTarget)) {
      target = {
        node: originalTarget
      };
    }
    this._target = target;
    this.emitter.emit('change', target);
  };
  _proto.onChange = function onChange(fn) {
    var _this = this;
    this.emitter.addListener('change', fn);
    return function () {
      _this.emitter.removeListener('change', fn);
    };
  };
  (0, _createClass2["default"])(ActiveTracker, [{
    key: "currentNode",
    get: function get() {
      var _this$_target;
      return (_this$_target = this._target) === null || _this$_target === void 0 ? void 0 : _this$_target.node;
    }
  }, {
    key: "detail",
    get: function get() {
      var _this$_target2;
      return (_this$_target2 = this._target) === null || _this$_target2 === void 0 ? void 0 : _this$_target2.detail;
    }

    /**
     * @deprecated
     */
    /* istanbul ignore next */
  }, {
    key: "intance",
    get: function get() {
      return this.instance;
    }
  }, {
    key: "instance",
    get: function get() {
      var _this$_target3;
      return (_this$_target3 = this._target) === null || _this$_target3 === void 0 ? void 0 : _this$_target3.instance;
    }
  }]);
  return ActiveTracker;
}(), (_descriptor = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "_target", [_dec], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
})), _class));
