// @alilc/lowcode-designer/lib/designer/designer-view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.DesignerView = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = require("react");
var _classnames = _interopRequireDefault(require("classnames"));
var _dragGhost = _interopRequireDefault(require("./drag-ghost"));
var _designer = require("./designer");
var _project = require("../project");
require("./designer.less");
var _excluded = ["designer"];
var DesignerView = exports.DesignerView = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(DesignerView, _Component);
  function DesignerView(props) {
    var _this;
    _this = _Component.call(this, props) || this;
    _this.designer = void 0;
    _this.viewName = void 0;
    var designer = props.designer,
      designerProps = (0, _objectWithoutPropertiesLoose2["default"])(props, _excluded);
    _this.viewName = designer === null || designer === void 0 ? void 0 : designer.viewName;
    if (designer) {
      _this.designer = designer;
      designer.setProps(designerProps);
    } else {
      _this.designer = new _designer.Designer(designerProps);
    }
    return _this;
  }
  var _proto = DesignerView.prototype;
  _proto.shouldComponentUpdate = function shouldComponentUpdate(nextProps) {
    this.designer.setProps(nextProps);
    var props = this.props;
    if (nextProps.className !== props.className || nextProps.style !== props.style || nextProps.dragGhostComponent !== props.dragGhostComponent) {
      return true;
    }
    return false;
  };
  _proto.componentDidMount = function componentDidMount() {
    var onMount = this.props.onMount;
    if (onMount) {
      onMount(this.designer);
    }
    this.designer.postEvent('mount', this.designer);
  };
  _proto.UNSAFE_componentWillMount = function UNSAFE_componentWillMount() {
    this.designer.purge();
  };
  _proto.render = function render() {
    var _this$props = this.props,
      className = _this$props.className,
      style = _this$props.style,
      dragGhostComponent = _this$props.dragGhostComponent;
    var DragGhost = dragGhostComponent || _dragGhost["default"];
    return /*#__PURE__*/React.createElement("div", {
      className: (0, _classnames["default"])('lc-designer', className),
      style: style
    }, /*#__PURE__*/React.createElement(DragGhost, {
      designer: this.designer
    }), /*#__PURE__*/React.createElement(_project.ProjectView, {
      designer: this.designer
    }));
  };
  return DesignerView;
}(_react.Component);
