// @alilc/lowcode-utils/lib/script.js
"use strict";

exports.__esModule = true;
exports.evaluate = evaluate;
exports.evaluateExpression = evaluateExpression;
exports.load = load;
exports.newFunction = newFunction;
var _createDefer = require("./create-defer");
var _logger = require("./logger");
var logger = new _logger.Logger({
  level: 'warn',
  bizName: 'utils'
});
function evaluate(script, scriptType) {
  var scriptEl = document.createElement('script');
  scriptType && (scriptEl.type = scriptType);
  scriptEl.text = script;
  document.head.appendChild(scriptEl);
  document.head.removeChild(scriptEl);
}
function load(url, scriptType) {
  var node = document.createElement('script');

  // node.setAttribute('crossorigin', 'anonymous');

  node.onload = onload;
  node.onerror = onload;
  var i = (0, _createDefer.createDefer)();
  function onload(e) {
    node.onload = null;
    node.onerror = null;
    if (e.type === 'load') {
      i.resolve();
    } else {
      i.reject();
    }
    // document.head.removeChild(node);
    // node = null;
  }
  node.src = url;

  // `async=false` is required to make sure all js resources execute sequentially.
  node.async = false;
  scriptType && (node.type = scriptType);
  document.head.appendChild(node);
  return i.promise();
}
function evaluateExpression(expr) {
  // eslint-disable-next-line no-new-func
  var fn = new Function(expr);
  return fn();
}
function newFunction(args, code) {
  try {
    // eslint-disable-next-line no-new-func
    return new Function(args, code);
  } catch (e) {
    logger.warn('Caught error, Cant init func');
    return null;
  }
}
