// @alilc/lowcode-utils/lib/transaction-manager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.transactionManager = void 0;
var _lowcodeTypes = require("@alilc/lowcode-types");
var _mobx = require("mobx");
var _events = _interopRequireDefault(require("events"));
var TransactionManager = function TransactionManager() {
  var _this = this;
  this.emitter = new _events["default"]();
  this.executeTransaction = function (fn, type) {
    if (type === void 0) {
      type = _lowcodeTypes.IPublicEnumTransitionType.REPAINT;
    }
    _this.emitter.emit("[" + type + "]startTransaction");
    (0, _mobx.runInAction)(fn);
    _this.emitter.emit("[" + type + "]endTransaction");
  };
  this.onStartTransaction = function (fn, type) {
    if (type === void 0) {
      type = _lowcodeTypes.IPublicEnumTransitionType.REPAINT;
    }
    _this.emitter.on("[" + type + "]startTransaction", fn);
    return function () {
      _this.emitter.off("[" + type + "]startTransaction", fn);
    };
  };
  this.onEndTransaction = function (fn, type) {
    if (type === void 0) {
      type = _lowcodeTypes.IPublicEnumTransitionType.REPAINT;
    }
    _this.emitter.on("[" + type + "]endTransaction", fn);
    return function () {
      _this.emitter.off("[" + type + "]endTransaction", fn);
    };
  };
};
var transactionManager = exports.transactionManager = new TransactionManager();
