// @alilc/lowcode-plugin-code-editor/lib/pane/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.CodeEditorPane = void 0;
var _dialog = _interopRequireDefault(require("@alifd/next/lib/dialog"));
var _message = _interopRequireDefault(require("@alifd/next/lib/message"));
var _react = _interopRequireWildcard(require("react"));
var _components = require("../components");
var _utils = require("../utils");
var _config = require("../config");
var _lowcodeEngine = require("@alilc/lowcode-engine");
require("@alilc/lowcode-plugin-base-monaco-editor/lib/style");
require("./index.less");
var _SaveIcon = require("../components/SaveIcon");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
var CodeEditorPane = exports.CodeEditorPane = /*#__PURE__*/(0, _react.memo)(function (_ref) {
  var project = _ref.project,
    event = _ref.event,
    skeleton = _ref.skeleton;
  var _useState = (0, _react.useState)(_config.TAB_KEY.JS),
    activeKey = _useState[0],
    setActiveKey = _useState[1];
  var lowcodeProjectRef = (0, _react.useRef)(project);
  var skeletonRef = (0, _react.useRef)(skeleton);
  var eventRef = (0, _react.useRef)(event);
  var jsEditorRef = (0, _react.useRef)(null);
  var cssEditorRef = (0, _react.useRef)(null);
  var saveSchemaRef = (0, _react.useRef)(); // save code to schema

  var _useState2 = (0, _react.useState)(function () {
      return project.exportSchema(_lowcodeEngine.common.designerCabin.TransformStage.Save);
    }),
    schema = _useState2[0],
    setSchema = _useState2[1];
  var jsCode = (0, _react.useMemo)(function () {
    return (0, _utils.schema2JsCode)(schema);
  }, [schema]);
  var cssCode = (0, _react.useMemo)(function () {
    return (0, _utils.schema2CssCode)(schema);
  }, [schema]);
  (0, _react.useEffect)(function () {
    saveSchemaRef.current = function () {
      try {
        var _lowcodeProjectRef$cu, _jsEditorRef$current$, _jsEditorRef$current, _cssEditorRef$current, _cssEditorRef$current2, _lowcodeProjectRef$cu2;
        var currentSchema = (_lowcodeProjectRef$cu = lowcodeProjectRef.current) === null || _lowcodeProjectRef$cu === void 0 ? void 0 : _lowcodeProjectRef$cu.exportSchema(_lowcodeEngine.common.designerCabin.TransformStage.Save);
        var pageNode = currentSchema.componentsTree[0];
        var _ref2 = (_jsEditorRef$current$ = (_jsEditorRef$current = jsEditorRef.current) === null || _jsEditorRef$current === void 0 ? void 0 : _jsEditorRef$current.getSchemaFromCode()) !== null && _jsEditorRef$current$ !== void 0 ? _jsEditorRef$current$ : {},
          state = _ref2.state,
          methods = _ref2.methods,
          lifeCycles = _ref2.lifeCycles,
          _ref2$originCode = _ref2.originCode,
          originCode = _ref2$originCode === void 0 ? '' : _ref2$originCode;
        var css = (_cssEditorRef$current = (_cssEditorRef$current2 = cssEditorRef.current) === null || _cssEditorRef$current2 === void 0 ? void 0 : _cssEditorRef$current2.getBeautifiedCSS()) !== null && _cssEditorRef$current !== void 0 ? _cssEditorRef$current : cssCode;
        pageNode.state = state;
        pageNode.methods = methods;
        pageNode.lifeCycles = lifeCycles;
        pageNode.originCode = originCode;
        pageNode.css = css;
        (_lowcodeProjectRef$cu2 = lowcodeProjectRef.current) === null || _lowcodeProjectRef$cu2 === void 0 ? void 0 : _lowcodeProjectRef$cu2.importSchema(currentSchema);
        setSchema(currentSchema);
        _message["default"].success({
          content: _config.WORDS.saveSuccess,
          duration: 1000
        });
      } catch (err) {
        if (err instanceof Error) {
          _dialog["default"].alert({
            title: _config.WORDS.title,
            content: /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, _config.WORDS.generalParseError, /*#__PURE__*/_react["default"].createElement("pre", null, err.message)),
            onOk: function onOk() {
              var _skeletonRef$current;
              (_skeletonRef$current = skeletonRef.current) === null || _skeletonRef$current === void 0 ? void 0 : _skeletonRef$current.showPanel('codeEditor');
            }
          });
        }
        // eslint-disable-next-line no-console
        console.error(err);
      }
    };
  }, [cssCode]);
  (0, _react.useEffect)(function () {
    lowcodeProjectRef.current = project;
  }, [project]);
  (0, _react.useEffect)(function () {
    skeletonRef.current = skeleton;
  }, [skeleton]);
  (0, _react.useEffect)(function () {
    eventRef.current = event;
  }, [event]);
  (0, _react.useEffect)(function () {
    var _skeletonRef$current2, _skeletonRef$current3, _eventRef$current, _eventRef$current2;
    // load schema on open
    (_skeletonRef$current2 = skeletonRef.current) === null || _skeletonRef$current2 === void 0 ? void 0 : _skeletonRef$current2.onShowPanel(function (pluginName) {
      if (pluginName === 'codeEditor') {
        var _lowcodeProjectRef$cu3, _jsEditorRef$current2, _cssEditorRef$current3;
        var _schema = (_lowcodeProjectRef$cu3 = lowcodeProjectRef.current) === null || _lowcodeProjectRef$cu3 === void 0 ? void 0 : _lowcodeProjectRef$cu3.exportSchema(_lowcodeEngine.common.designerCabin.TransformStage.Save);
        if (!_schema) {
          return;
        }
        var _jsCode = (0, _utils.schema2JsCode)(_schema);
        var _cssCode = (0, _utils.schema2CssCode)(_schema);
        setSchema(_schema);
        (_jsEditorRef$current2 = jsEditorRef.current) === null || _jsEditorRef$current2 === void 0 ? void 0 : _jsEditorRef$current2._updateCode(_jsCode);
        (_cssEditorRef$current3 = cssEditorRef.current) === null || _cssEditorRef$current3 === void 0 ? void 0 : _cssEditorRef$current3._updateCode(_cssCode);
      }
    });

    // save schema when panel closed
    (_skeletonRef$current3 = skeletonRef.current) === null || _skeletonRef$current3 === void 0 ? void 0 : _skeletonRef$current3.onHidePanel(function (pluginName) {
      if (pluginName === 'codeEditor') {
        var _saveSchemaRef$curren;
        (_saveSchemaRef$curren = saveSchemaRef.current) === null || _saveSchemaRef$curren === void 0 ? void 0 : _saveSchemaRef$curren.call(saveSchemaRef);
      }
    });

    // focus function by functionName
    (_eventRef$current = eventRef.current) === null || _eventRef$current === void 0 ? void 0 : _eventRef$current.on('common:codeEditor.focusByFunction', function (params) {
      setActiveKey(_config.TAB_KEY.JS);
      setTimeout(function () {
        var _jsEditorRef$current3;
        (_jsEditorRef$current3 = jsEditorRef.current) === null || _jsEditorRef$current3 === void 0 ? void 0 : _jsEditorRef$current3.focusByFunctionName(params);
      }, 100);
    });
    (_eventRef$current2 = eventRef.current) === null || _eventRef$current2 === void 0 ? void 0 : _eventRef$current2.on('common:codeEditor.addFunction', function (params) {
      setActiveKey(_config.TAB_KEY.JS);
      setTimeout(function () {
        var _jsEditorRef$current4;
        (_jsEditorRef$current4 = jsEditorRef.current) === null || _jsEditorRef$current4 === void 0 ? void 0 : _jsEditorRef$current4.addFunction(params);
      }, 100);
    });
  }, []);
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: "plugin-code-editor-pane"
  }, /*#__PURE__*/_react["default"].createElement(_components.JsEditor, {
    jsCode: jsCode,
    ref: jsEditorRef,
    currentTab: activeKey,
    onTabChange: function onTabChange(key) {
      return setActiveKey(key);
    }
  }), /*#__PURE__*/_react["default"].createElement(_components.CssEditor, {
    cssCode: cssCode,
    ref: cssEditorRef,
    currentTab: activeKey,
    onTabChange: function onTabChange(key) {
      return setActiveKey(key);
    }
  }), /*#__PURE__*/_react["default"].createElement(_SaveIcon.SaveIcon, {
    onClick: function onClick() {
      var _saveSchemaRef$curren2;
      (_saveSchemaRef$curren2 = saveSchemaRef.current) === null || _saveSchemaRef$curren2 === void 0 ? void 0 : _saveSchemaRef$curren2.call(saveSchemaRef);
    }
    // isDisabled={code === jsCode || hasError}
  }));
});

CodeEditorPane.displayName = 'LowcodeCodeEditor';
