// @alilc/lowcode-engine-ext/lib/setter/style-setter/pro/position/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports["default"] = void 0;
var _numberPicker = _interopRequireDefault(require("@alifd/next/lib/number-picker"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _select = _interopRequireDefault(require("@alifd/next/lib/select"));
var _collapse = _interopRequireDefault(require("@alifd/next/lib/collapse"));
var React = _interopRequireWildcard(require("react"));
var _row = _interopRequireDefault(require("../../components/row"));
var _positionBox = _interopRequireDefault(require("../position/positionBox"));
var _config = _interopRequireDefault(require("./config.json"));
var _utils = require("../../utils");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var Panel = _collapse["default"].Panel;
var _default = function _default(props) {
  var _float = _config["default"]["float"],
    clear = _config["default"].clear,
    position = _config["default"].position;
  var onStyleChange = props.onStyleChange,
    styleData = props.styleData,
    positionPropsConfig = props.positionPropsConfig;
  var isShowFloat = positionPropsConfig.isShowFloat,
    isShowClear = positionPropsConfig.isShowClear;
  var onZIndexChange = function onZIndexChange(zIndex) {
    onStyleChange([{
      styleKey: 'zIndex',
      value: zIndex
    }]);
  };
  var initData = function initData() {};
  (0, React.useEffect)(function () {
    initData();
  }, []);
  return /*#__PURE__*/React.createElement(_collapse["default"], {
    defaultExpandedKeys: ['0']
  }, /*#__PURE__*/React.createElement(Panel, {
    title: "\u4F4D\u7F6E"
  }, /*#__PURE__*/React.createElement(_row["default"], {
    title: position.title,
    styleData: styleData,
    styleKey: "position"
  }, /*#__PURE__*/React.createElement(_select["default"], {
    dataSource: position.dataList,
    value: styleData.position,
    hasClear: true,
    onChange: function onChange(val) {
      return onStyleChange([{
        styleKey: 'position',
        value: val
      }]);
    }
  })), styleData.position && styleData.position != 'static' && /*#__PURE__*/React.createElement(_positionBox["default"], (0, _extends2["default"])({
    styleData: styleData,
    onStyleChange: onStyleChange
  }, props)), /*#__PURE__*/React.createElement(_row["default"], {
    title: 'zIndex',
    styleData: styleData,
    styleKey: "zIndex"
  }, /*#__PURE__*/React.createElement(_numberPicker["default"], {
    disabled: (0, _utils.isCssVarBind)(styleData.zIndex),
    step: 1,
    precision: 2,
    onChange: onZIndexChange,
    value: styleData.zIndex
  })), isShowFloat && /*#__PURE__*/React.createElement(_row["default"], {
    title: _float.title,
    dataList: _float.dataList,
    onStyleChange: onStyleChange,
    styleData: styleData,
    styleKey: "float"
  }), isShowClear && /*#__PURE__*/React.createElement(_row["default"], {
    title: clear.title,
    dataList: clear.dataList,
    onStyleChange: onStyleChange,
    styleData: styleData,
    styleKey: "clear"
  })));
};
exports["default"] = _default;
