// @alilc/lowcode-workspace/lib/view/resource-view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.ResourceView = void 0;
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = require("react");
var _editorView = require("./editor-view");
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
require("./resource-view.less");
var _lowcodeEditorSkeleton = require("@alilc/lowcode-editor-skeleton");
var _class;
var ResourceView = exports.ResourceView = (0, _lowcodeEditorCore.observer)(_class = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(ResourceView, _PureComponent);
  function ResourceView() {
    return _PureComponent.apply(this, arguments) || this;
  }
  var _proto = ResourceView.prototype;
  _proto.render = function render() {
    var skeleton = this.props.resource.skeleton;
    var editorViews = this.props.window.editorViews;
    return /*#__PURE__*/React.createElement("div", {
      className: "workspace-resource-view"
    }, /*#__PURE__*/React.createElement(_lowcodeEditorSkeleton.TopArea, {
      area: skeleton.topArea,
      itemClassName: "engine-actionitem"
    }), /*#__PURE__*/React.createElement("div", {
      className: "workspace-editor-body"
    }, Array.from(editorViews.values()).map(function (editorView) {
      return /*#__PURE__*/React.createElement(_editorView.EditorView, {
        key: editorView.name,
        active: editorView.active,
        editorView: editorView
      });
    })));
  };
  return ResourceView;
}(_react.PureComponent)) || _class;
