// @alilc/lowcode-workspace/lib/view/window-view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.WindowView = void 0;
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = require("react");
var _resourceView = require("./resource-view");
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _lowcodeDesigner = require("@alilc/lowcode-designer");
var _webview = require("../inner-plugins/webview");
var _class;
var WindowView = exports.WindowView = (0, _lowcodeEditorCore.observer)(_class = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(WindowView, _PureComponent);
  function WindowView() {
    return _PureComponent.apply(this, arguments) || this;
  }
  var _proto = WindowView.prototype;
  _proto.render = function render() {
    var active = this.props.active;
    var _this$props$window = this.props.window,
      resource = _this$props$window.resource,
      initReady = _this$props$window.initReady,
      url = _this$props$window.url;
    if (!initReady) {
      var Loading = _lowcodeEditorCore.engineConfig.get('loadingComponent', _lowcodeDesigner.BuiltinLoading);
      return /*#__PURE__*/React.createElement("div", {
        className: "workspace-engine-main " + (active ? 'active' : '')
      }, /*#__PURE__*/React.createElement(Loading, null));
    }
    if (resource.type === 'webview' && url) {
      return /*#__PURE__*/React.createElement(_webview.DesignerView, {
        url: url,
        viewName: resource.name
      });
    }
    return /*#__PURE__*/React.createElement("div", {
      className: "workspace-engine-main " + (active ? 'active' : '')
    }, /*#__PURE__*/React.createElement(_resourceView.ResourceView, {
      resource: resource,
      window: this.props.window
    }));
  };
  return WindowView;
}(_react.PureComponent)) || _class;
