// universal-request/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = _default;

var _universalEnv = require("universal-env");

var _types = require("./types");

var _utils = require("./utils");

var _index = _interopRequireDefault(require("./web/index"));

var _index2 = _interopRequireDefault(require("./weex/index"));

var _index3 = _interopRequireDefault(require("./miniapp/index"));

var _index4 = _interopRequireDefault(require("./wechat-miniprogram/index"));

var _index5 = _interopRequireDefault(require("./miniapp-bytedance/index"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function dutyChain() {
  for (var _len = arguments.length, fns = new Array(_len), _key = 0; _key < _len; _key++) {
    fns[_key] = arguments[_key];
  }

  for (var i = 0; i < fns.length; i++) {
    var result = fns[i]();

    if (result) {
      return result;
    }
  }
}

function handleWeb(afterOptions) {
  if (_universalEnv.isWeb) {
    return (0, _index.default)(afterOptions);
  }

  return null;
}

function handleMiniApp(afterOptions) {
  if (_universalEnv.isMiniApp) {
    return (0, _index3.default)(afterOptions);
  }

  return null;
}

function handleWeex(afterOptions) {
  if (_universalEnv.isWeex) {
    return (0, _index2.default)(afterOptions);
  }

  return null;
}

function handleWeChatMiniprogram(afterOptions) {
  if (_universalEnv.isWeChatMiniProgram) {
    return (0, _index4.default)(afterOptions);
  }

  return null;
}

function handleBytedanceMiniprogram(afterOptions) {
  if (_universalEnv.isByteDanceMicroApp) {
    return (0, _index5.default)(afterOptions);
  }
}

function _default(options) {
  var afterOptions = Object.assign({}, _types.DEFAULT_REQUEST_OPTIONS, options, {
    method: (options.method || 'GET').toUpperCase()
  });
  options.headers = (0, _utils.normalizeHeaders)(options.headers || {});
  return dutyChain(handleWeex.bind(null, afterOptions), handleWeb.bind(null, afterOptions), handleMiniApp.bind(null, afterOptions), handleBytedanceMiniprogram.bind(null, afterOptions), handleWeChatMiniprogram.bind(null, afterOptions));
}
