// universal-request/lib/types.js
"use strict";

exports.__esModule = true;
exports.ERROR_REQUEST_TIMEOUT = exports.ERROR_REQUEST_NOT_SUPPORT = exports.ERROR_REQUEST_ABORT = exports.DEFAULT_TIMEOUT = exports.DEFAULT_REQUEST_OPTIONS = exports.DEFAULT_JSON_CALLBACK = exports.DEFAULT_JSONP_OPTIONS = exports.DATA_TYPE = void 0;

/**
 *  Common
 */
var DATA_TYPE;
exports.DATA_TYPE = DATA_TYPE;

(function (DATA_TYPE) {
  DATA_TYPE["json"] = "json";
  DATA_TYPE["text"] = "text";
})(DATA_TYPE || (exports.DATA_TYPE = DATA_TYPE = {}));

var DEFAULT_TIMEOUT = 20000;
exports.DEFAULT_TIMEOUT = DEFAULT_TIMEOUT;
var DEFAULT_JSON_CALLBACK = 'callback';
exports.DEFAULT_JSON_CALLBACK = DEFAULT_JSON_CALLBACK;
var DEFAULT_JSONP_OPTIONS = {
  timeout: DEFAULT_TIMEOUT,
  jsonpCallback: DEFAULT_JSON_CALLBACK,
  dataType: DATA_TYPE.json
};
exports.DEFAULT_JSONP_OPTIONS = DEFAULT_JSONP_OPTIONS;
var DEFAULT_REQUEST_OPTIONS = {
  url: '',
  headers: {
    'Content-Type': 'application/json'
  },
  method: 'GET',
  timeout: DEFAULT_TIMEOUT,
  dataType: DATA_TYPE.json
};
exports.DEFAULT_REQUEST_OPTIONS = DEFAULT_REQUEST_OPTIONS;
var ERROR_REQUEST_ABORT = {
  code: 0,
  message: 'Request Abort'
};
exports.ERROR_REQUEST_ABORT = ERROR_REQUEST_ABORT;
var ERROR_REQUEST_TIMEOUT = {
  code: 1,
  message: 'Request timeout'
};
exports.ERROR_REQUEST_TIMEOUT = ERROR_REQUEST_TIMEOUT;
var ERROR_REQUEST_NOT_SUPPORT = {
  code: 2,
  message: 'Request not support this platform'
};
exports.ERROR_REQUEST_NOT_SUPPORT = ERROR_REQUEST_NOT_SUPPORT;
