// universal-request/lib/utils.js
"use strict";

exports.__esModule = true;
exports.applyParamToURL = applyParamToURL;
exports.checkIsApplyDataToURL = checkIsApplyDataToURL;
exports.getDataWithType = getDataWithType;
exports.isObject = isObject;
exports.isPlainObject = isPlainObject;
exports.normalizeHeaders = normalizeHeaders;
exports.object2json = object2json;
exports.stringifyQS = stringifyQS;

var _types = require("./types");

function getDataWithType(data, type) {
  if (type === _types.DATA_TYPE.json) {
    try {
      return JSON.parse(data);
    } catch (e) {}
  }

  if (type === _types.DATA_TYPE.text) {
    try {
      return JSON.stringify(data);
    } catch (e) {}
  }

  return data;
}

function stringifyQS(qs) {
  if (!qs) {
    return '';
  }

  var str = [];

  for (var key in qs) {
    if (qs[key] !== undefined && qs[key] !== null) {
      str.push(key + "=" + encodeURIComponent(String(qs[key])));
    }
  }

  return str.join('&');
}

function applyParamToURL(param, url) {
  if (!param) {
    return url;
  }

  return "" + url + (url.indexOf('?') === -1 ? '?' : '&') + stringifyQS(param);
}

function object2json(obj) {
  try {
    return JSON.stringify(obj);
  } catch (e) {
    return obj || '';
  }
}

function isObject(obj) {
  return typeof obj === 'object' && obj !== null;
}

function normalizeHeaders(obj) {
  if (!isObject(obj)) {
    return obj;
  }

  ['Accept', 'Content-Type'].forEach(function (key) {
    Object.keys(obj).some(function (headerKey) {
      if (headerKey.toUpperCase() === key.toUpperCase() && headerKey !== key) {
        obj[key] = obj[headerKey];
        delete obj[headerKey];
        return true;
      }

      return false;
    });
  });
  return obj;
}

function checkIsApplyDataToURL(headers) {
  if (!headers || !headers['Content-Type']) {
    return false;
  }

  return String(headers['Content-Type']).toLowerCase().indexOf('application/x-www-form-urlencoded') > -1;
}

var EMPTY_OBJECT = {};

function isPlainObject(obj) {
  return EMPTY_OBJECT.toString.call(obj) === '[object Object]';
}
