// @alilc/lowcode-shell/lib/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  Detecting: true,
  DocumentModel: true,
  History: true,
  Node: true,
  NodeChildren: true,
  Prop: true,
  Selection: true,
  Dragon: true,
  SettingTopEntry: true,
  Clipboard: true,
  SettingPropEntry: true,
  SettingField: true,
  Window: true,
  SkeletonItem: true,
  Project: true,
  Material: true,
  Logger: true,
  Plugins: true,
  Skeleton: true,
  Setters: true,
  Hotkey: true,
  Common: true,
  getEvent: true,
  Event: true,
  Canvas: true,
  Workspace: true,
  SimulatorHost: true,
  Config: true,
  CommonUI: true,
  Command: true
};
var _model = require("./model");
exports.Detecting = _model.Detecting;
exports.DocumentModel = _model.DocumentModel;
exports.History = _model.History;
exports.Node = _model.Node;
exports.NodeChildren = _model.NodeChildren;
exports.Prop = _model.Prop;
exports.Selection = _model.Selection;
exports.Dragon = _model.Dragon;
exports.SettingTopEntry = _model.SettingTopEntry;
exports.Clipboard = _model.Clipboard;
exports.SettingPropEntry = _model.SettingField;
exports.SettingField = _model.SettingField;
exports.Window = _model.Window;
exports.SkeletonItem = _model.SkeletonItem;
var _api = require("./api");
exports.Project = _api.Project;
exports.Material = _api.Material;
exports.Logger = _api.Logger;
exports.Plugins = _api.Plugins;
exports.Skeleton = _api.Skeleton;
exports.Setters = _api.Setters;
exports.Hotkey = _api.Hotkey;
exports.Common = _api.Common;
exports.getEvent = _api.getEvent;
exports.Event = _api.Event;
exports.Canvas = _api.Canvas;
exports.Workspace = _api.Workspace;
exports.SimulatorHost = _api.SimulatorHost;
exports.Config = _api.Config;
exports.CommonUI = _api.CommonUI;
exports.Command = _api.Command;
var _symbols = require("./symbols");
Object.keys(_symbols).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _symbols[key]) return;
  exports[key] = _symbols[key];
});
