// @alilc/lowcode-shell/lib/symbols.js
"use strict";

exports.__esModule = true;
exports.workspaceSymbol = exports.windowSymbol = exports.skeletonSymbol = exports.skeletonItemSymbol = exports.skeletonCabinSymbol = exports.simulatorRenderSymbol = exports.simulatorHostSymbol = exports.settingTopEntrySymbol = exports.settingFieldSymbol = exports.selectionSymbol = exports.resourceTypeSymbol = exports.resourceSymbol = exports.propsSymbol = exports.propSymbol = exports.projectSymbol = exports.pluginsSymbol = exports.pluginInstanceSymbol = exports.pluginContextSymbol = exports.nodeSymbol = exports.nodeChildrenSymbol = exports.modalNodesManagerSymbol = exports.locateEventSymbol = exports.hotkeySymbol = exports.historySymbol = exports.editorViewSymbol = exports.editorSymbol = exports.editorCabinSymbol = exports.dropLocationSymbol = exports.dragonSymbol = exports.dragObjectSymbol = exports.documentSymbol = exports.detectingSymbol = exports.designerSymbol = exports.designerCabinSymbol = exports.configSymbol = exports.conditionGroupSymbol = exports.componentMetaSymbol = exports.commandSymbol = exports.clipboardSymbol = exports.canvasSymbol = void 0;
/**
 * 以下 symbol 均用于在 shell 层对外暴露的模型中存储相应内部模型的 key
 */
var projectSymbol = exports.projectSymbol = Symbol('project');
var designerSymbol = exports.designerSymbol = Symbol('designer');
var skeletonSymbol = exports.skeletonSymbol = Symbol('skeleton');
var documentSymbol = exports.documentSymbol = Symbol('document');
var editorSymbol = exports.editorSymbol = Symbol('editor');
var nodeSymbol = exports.nodeSymbol = Symbol('node');
var modalNodesManagerSymbol = exports.modalNodesManagerSymbol = Symbol('modalNodesManager');
var nodeChildrenSymbol = exports.nodeChildrenSymbol = Symbol('nodeChildren');
var propSymbol = exports.propSymbol = Symbol('prop');
var settingFieldSymbol = exports.settingFieldSymbol = Symbol('settingField');
var settingTopEntrySymbol = exports.settingTopEntrySymbol = Symbol('settingTopEntry');
var propsSymbol = exports.propsSymbol = Symbol('props');
var detectingSymbol = exports.detectingSymbol = Symbol('detecting');
var selectionSymbol = exports.selectionSymbol = Symbol('selection');
var historySymbol = exports.historySymbol = Symbol('history');
var canvasSymbol = exports.canvasSymbol = Symbol('canvas');
var dragonSymbol = exports.dragonSymbol = Symbol('dragon');
var componentMetaSymbol = exports.componentMetaSymbol = Symbol('componentMeta');
var dropLocationSymbol = exports.dropLocationSymbol = Symbol('dropLocation');
var simulatorHostSymbol = exports.simulatorHostSymbol = Symbol('simulatorHost');
var simulatorRenderSymbol = exports.simulatorRenderSymbol = Symbol('simulatorRender');
var dragObjectSymbol = exports.dragObjectSymbol = Symbol('dragObject');
var locateEventSymbol = exports.locateEventSymbol = Symbol('locateEvent');
var designerCabinSymbol = exports.designerCabinSymbol = Symbol('designerCabin');
var editorCabinSymbol = exports.editorCabinSymbol = Symbol('editorCabin');
var skeletonCabinSymbol = exports.skeletonCabinSymbol = Symbol('skeletonCabin');
var hotkeySymbol = exports.hotkeySymbol = Symbol('hotkey');
var pluginsSymbol = exports.pluginsSymbol = Symbol('plugins');
var workspaceSymbol = exports.workspaceSymbol = Symbol('workspace');
var windowSymbol = exports.windowSymbol = Symbol('window');
var pluginInstanceSymbol = exports.pluginInstanceSymbol = Symbol('plugin-instance');
var resourceTypeSymbol = exports.resourceTypeSymbol = Symbol('resourceType');
var resourceSymbol = exports.resourceSymbol = Symbol('resource');
var clipboardSymbol = exports.clipboardSymbol = Symbol('clipboard');
var configSymbol = exports.configSymbol = Symbol('configSymbol');
var conditionGroupSymbol = exports.conditionGroupSymbol = Symbol('conditionGroup');
var editorViewSymbol = exports.editorViewSymbol = Symbol('editorView');
var pluginContextSymbol = exports.pluginContextSymbol = Symbol('pluginContext');
var skeletonItemSymbol = exports.skeletonItemSymbol = Symbol('skeletonItem');
var commandSymbol = exports.commandSymbol = Symbol('command');
