// @alifd/meet-react/lib/select/section/select-search.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
var _babelRuntimeJsxStyleTransform = require("@alifd/babel-runtime-jsx-style-transform");
var _react = _interopRequireWildcard(require("react"));
var _search = _interopRequireDefault(require("../../search"));
var _utils = require("../../utils");
var _view = _interopRequireDefault(require("../../view"));
var _context = _interopRequireDefault(require("../context"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var SelectSearch = function (props, ref) {
  var _useContext = (0, _react.useContext)(_context.default),
    prefix = _useContext.prefix,
    locale = _useContext.locale,
    setSearchValue = _useContext.setSearchValue,
    searchValue = _useContext.value,
    innerValue = _useContext.innerValue,
    contextSearch = _useContext.onSearch;
  var onSearch = props.onSearch;
  var _useState = (0, _react.useState)(''),
    value = _useState[0],
    _setValue = _useState[1];
  (0, _react.useImperativeHandle)(ref, function () {
    return {
      setValue: function setValue(v) {
        _setValue(v);
        setSearchValue(v);
      }
    };
  });
  return /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: prefix + "select-search-wrapper",
    style: (0, _babelRuntimeJsxStyleTransform.rpx2vw4style)({
      width: '100%',
      height: '100rpx'
    })
  }, /*#__PURE__*/(0, _react.createElement)(_search.default, {
    placeholder: locale.searchTipText,
    value: value,
    hasClear: true,
    onChange: function handleSearch(v) {
      _setValue(v);
      setSearchValue(v);
      if ((0, _utils.isFunction)(onSearch)) {
        onSearch(v);
      }
      if ((0, _utils.isFunction)(contextSearch)) {
        contextSearch(v, searchValue, innerValue);
      }
    }
  }));
};
var _default = exports.default = /*#__PURE__*/(0, _react.forwardRef)(SelectSearch);
