// @alifd/meet-react/lib/select/section/select-value.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _classnames = _interopRequireDefault(require("classnames"));
var _form = require("../../form");
var _tagSelect = _interopRequireDefault(require("../../tag-select"));
var _view = _interopRequireDefault(require("../../view"));
var _tag = _interopRequireDefault(require("../../tag"));
var _hooks = require("../../utils/hooks");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var SelectValue = function (props) {
  var _classNames, _classNames2;
  var items = props.items,
    prefix = props.prefix,
    displayType = props.displayType,
    sizeProp = props.size,
    disabled = props.disabled,
    type = props.type;
  var _useContext = (0, _react.useContext)(_form.ItemContext),
    isInsideForm = _useContext.isInsideForm,
    labelAlign = _useContext.labelAlign,
    contextSize = _useContext.size;
  var clsPrefix = prefix + "select";
  var isInverse = type === 'inverse';
  var size = sizeProp || (isInsideForm ? contextSize : 'medium');
  var ref = (0, _react.useRef)((0, _hooks.guid)('select-val-'));
  switch (displayType) {
    case 'tag':
      return /*#__PURE__*/(0, _react.createElement)(_tagSelect.default, {
        disabled: disabled,
        isPreview: true,
        inlineMode: displayType === 'tag' && isInsideForm && labelAlign === 'inset',
        prefix: prefix,
        size: size,
        dataSource: items
      });
    case 'tag-value':
      return /*#__PURE__*/(0, _react.createElement)(_view.default, {
        className: (0, _classnames.default)(clsPrefix + "-tag-value", (_classNames = {}, _classNames[clsPrefix + "-value--inverse"] = isInverse, _classNames[clsPrefix + "-value--disabled"] = disabled, _classNames[clsPrefix + "-value--inverse-disabled"] = disabled && isInverse, _classNames))
      }, items.map(function (val) {
        return /*#__PURE__*/(0, _react.createElement)(_tag.default, {
          key: ref.current + "-" + val,
          className: clsPrefix + "-tag-item",
          type: "primary",
          size: size
        }, "" + val);
      }));
    default:
      return /*#__PURE__*/(0, _react.createElement)(_view.default, {
        className: (0, _classnames.default)((_classNames2 = {}, _classNames2[clsPrefix + "-value--inverse"] = isInverse, _classNames2[clsPrefix + "-value--disabled"] = disabled, _classNames2[clsPrefix + "-value--inverse-disabled"] = disabled && isInverse, _classNames2))
      }, items.join(','));
  }
};
var _default = exports.default = SelectValue;
