// @alilc/lowcode-designer/lib/builtin-simulator/resource-consumer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports["default"] = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));
var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));
var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _renderer = require("./renderer");
var _dec, _class, _descriptor;
var UNSET = Symbol('unset');
// master 进程
//  0. 初始化该对象，因为需要响应变更发生在 master 进程
//  1. 提供消费数据或数据提供器，比如 Asset 资源，如果不是数据提供器，会持续提供
//  2. 收到成功通知
// renderer 进程
//  1. 持续消费，并持续监听数据
//  2. 消费
// 这里涉及俩个自定义项
//  1. 被消费数据协议
//  2. 消费机制（渲染进程自定 + 传递进入）
var ResourceConsumer = exports["default"] = (_dec = _lowcodeEditorCore.obx.ref, (_class = /*#__PURE__*/function () {
  function ResourceConsumer(provider, consumer) {
    var _this = this;
    this.consumer = consumer;
    this.emitter = (0, _lowcodeEditorCore.createModuleEventBus)('ResourceConsumer');
    (0, _initializerDefineProperty2["default"])(this, "_data", _descriptor, this);
    this._providing = void 0;
    this._consuming = void 0;
    this._firstConsumed = false;
    this.resolveFirst = void 0;
    (0, _lowcodeEditorCore.makeObservable)(this);
    this._providing = (0, _lowcodeEditorCore.autorun)(function () {
      _this._data = provider();
    });
  }
  var _proto = ResourceConsumer.prototype;
  _proto.consume = function consume(consumerOrRenderer) {
    var _this2 = this;
    if (this._consuming) {
      return;
    }
    var consumer;
    if ((0, _renderer.isSimulatorRenderer)(consumerOrRenderer)) {
      if (!this.consumer) {
        // TODO: throw error
        return;
      }
      var rendererConsumer = this.consumer;
      consumer = function consumer(data) {
        return rendererConsumer(consumerOrRenderer, data);
      };
    } else {
      consumer = consumerOrRenderer;
    }
    this._consuming = (0, _lowcodeEditorCore.autorun)( /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            if (!(_this2._data === UNSET)) {
              _context.next = 2;
              break;
            }
            return _context.abrupt("return");
          case 2:
            _context.next = 4;
            return consumer(_this2._data);
          case 4:
            // TODO: catch error and report
            if (_this2.resolveFirst) {
              _this2.resolveFirst();
            } else {
              _this2._firstConsumed = true;
            }
          case 5:
          case "end":
            return _context.stop();
        }
      }, _callee);
    })));
  };
  _proto.dispose = function dispose() {
    if (this._providing) {
      this._providing();
    }
    if (this._consuming) {
      this._consuming();
    }
    this.emitter.removeAllListeners();
  };
  _proto.waitFirstConsume = function waitFirstConsume() {
    var _this3 = this;
    if (this._firstConsumed) {
      return Promise.resolve();
    }
    return new Promise(function (resolve) {
      _this3.resolveFirst = resolve;
    });
  };
  return ResourceConsumer;
}(), (_descriptor = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "_data", [_dec], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return UNSET;
  }
})), _class));
