// @alilc/lowcode-engine-ext/lib/setter/style-setter/components/radio-group/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports["default"] = void 0;
var _balloon = _interopRequireDefault(require("@alifd/next/lib/balloon"));
var _radio = _interopRequireDefault(require("@alifd/next/lib/radio"));
var React = _interopRequireWildcard(require("react"));
var _icon = _interopRequireDefault(require("../icon"));
require("./index.less");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var RadioGroup = _radio["default"].Group;
var _default = function _default(props) {
  var dataList = props.dataList,
    styleKey = props.styleKey,
    onStyleChange = props.onStyleChange,
    value = props.value;
  var onRadioChange = function onRadioChange(key, val) {
    onStyleChange([{
      styleKey: key,
      value: val
    }]);
  };
  var onRadioItemClick = function onRadioItemClick(key, val, value) {
    if (value == val) {
      onStyleChange([{
        styleKey: key,
        value: null
      }]);
    }
  };
  return /*#__PURE__*/React.createElement("div", {
    className: "radiogroup-style"
  }, /*#__PURE__*/React.createElement(RadioGroup, {
    value: value,
    shape: "button",
    onChange: function onChange(val) {
      return onRadioChange(styleKey, val);
    }
    // onClick={(event) => {
    //   onRadioClick(event, styleKey, value);
    // }}
    ,
    "aria-labelledby": "groupId"
  }, dataList && dataList.map(function (item) {
    return /*#__PURE__*/React.createElement(_balloon["default"], {
      trigger: /*#__PURE__*/React.createElement(_radio["default"], {
        id: item.value,
        value: item.value,
        onClick: function onClick(e) {
          onRadioItemClick(styleKey, e.currentTarget.id, value);
        }
      }, item.icon ? /*#__PURE__*/React.createElement(_icon["default"], {
        type: item.icon,
        size: "small"
      }) : item.title),
      triggerType: "hover",
      closable: false,
      align: "t"
    }, item.tips);
  })));
};
exports["default"] = _default;
