// @alilc/lowcode-plugin-schema/es/editor.js
import _Button from "@alifd/next/es/button";
import _extends from "@babel/runtime/helpers/extends";
import _Message from "@alifd/next/es/message";
import _Dialog from "@alifd/next/es/dialog";
import * as React from 'react';
import { useState, useRef, useCallback, useEffect, useLayoutEffect } from 'react';
import MonacoEditor from '@alilc/lowcode-plugin-base-monaco-editor';
import { IPublicEnumTransformStage } from '@alilc/lowcode-types';
export default function PluginSchema(_ref) {
  var pluginContext = _ref.pluginContext,
    _ref$showProjectSchem = _ref.showProjectSchema,
    showProjectSchema = _ref$showProjectSchem === void 0 ? false : _ref$showProjectSchem;
  var project = pluginContext.project,
    skeleton = pluginContext.skeleton;
  var _useState = useState({
      width: 0,
      height: 0
    }),
    editorSize = _useState[0],
    setEditorSize = _useState[1];
  var getSchemaStr = useCallback(function () {
    var _schema$componentsTre;
    var schema = project.exportSchema(IPublicEnumTransformStage.Save);
    var schemaToShow = showProjectSchema ? schema : schema === null || schema === void 0 ? void 0 : (_schema$componentsTre = schema.componentsTree) === null || _schema$componentsTre === void 0 ? void 0 : _schema$componentsTre[0];
    return schemaToShow ? JSON.stringify(schemaToShow, null, 2) : '';
  }, []);
  var _useState2 = useState(function () {
      return getSchemaStr();
    }),
    schemaValue = _useState2[0],
    setSchemaValue = _useState2[1];
  var monacoEditorRef = useRef();
  var resize = useCallback(function () {
    setEditorSize({
      width: document.documentElement.clientWidth - 60,
      height: document.documentElement.clientHeight - 100
    });
  }, []);
  useLayoutEffect(function () {
    var cancelListenShowPanel = skeleton.onShowPanel(function (pluginName) {
      if (pluginName == 'LowcodePluginAliLowcodePluginSchema') {
        setSchemaValue(getSchemaStr());
      }
    });
    return cancelListenShowPanel;
  }, []);
  useEffect(function () {
    resize();
    window.addEventListener('resize', resize);
    return function () {
      window.removeEventListener('resize', resize);
    };
  }, [resize]);
  var onSave = function onSave() {
    _Dialog.alert({
      content: 'Are you 100% sure? Lowcode editor may crash.',
      footerActions: ['cancel', 'ok'],
      onOk: function onOk() {
        var json;
        try {
          var _monacoEditorRef$curr, _monacoEditorRef$curr2;
          json = JSON.parse((_monacoEditorRef$curr = (_monacoEditorRef$curr2 = monacoEditorRef.current) === null || _monacoEditorRef$curr2 === void 0 ? void 0 : _monacoEditorRef$curr2.getValue()) !== null && _monacoEditorRef$curr !== void 0 ? _monacoEditorRef$curr : schemaValue);
        } catch (err) {
          _Message.error('Cannot save schema. Schema Parse Error.' + err.message);
          return;
        }
        if (showProjectSchema) {
          // 当前操作项目级 schema
          project.importSchema(json);
        } else {
          // 当前操作页面级 schema
          project.importSchema(_extends({}, project.exportSchema(IPublicEnumTransformStage.Save), {
            componentsTree: [json]
          }));
        }
        _Message.success('Schema Saved!');
        skeleton.hidePanel('LowcodePluginAliLowcodePluginSchema');
      }
    });
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(_Button, {
    onClick: onSave,
    style: {
      position: 'absolute',
      right: 68,
      zIndex: 100,
      top: -38
    }
  }, pluginContext.intl('Save Schema')), /*#__PURE__*/React.createElement(MonacoEditor, {
    height: editorSize.height,
    language: "json",
    theme: "vs-light",
    value: schemaValue,
    onChange: function onChange(input) {
      setSchemaValue(input);
    },
    editorDidMount: function editorDidMount(_, monacoEditor) {
      monacoEditorRef.current = monacoEditor;
      monacoEditor.addAction({
        id: 'my-unique-id',
        label: 'Save Schema',
        keybindingContext: null,
        contextMenuGroupId: 'navigation',
        contextMenuOrder: 1.5,
        run: onSave
      });
    }
  }));
}
