// @alifd/meet-react/lib/select/option.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
var _tslib = require("tslib");
var _classnames = _interopRequireDefault(require("classnames"));
var _react = _interopRequireWildcard(require("react"));
var _checkbox = _interopRequireDefault(require("../checkbox"));
var _list = _interopRequireDefault(require("../list"));
var _context = _interopRequireDefault(require("./context"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
var Option = function (props) {
  var _classNames;
  var _props$disabled = props.disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled,
    children = props.children,
    className = props.className,
    value = props.value,
    _props$isNew = props.isNew,
    isNew = _props$isNew === void 0 ? false : _props$isNew,
    _props$onClick = props.onClick,
    _onClick = _props$onClick === void 0 ? function () {} : _props$onClick,
    _props$selected = props.selected,
    selected = _props$selected === void 0 ? false : _props$selected,
    title = props.title,
    others = (0, _tslib.__rest)(props, ["disabled", "children", "className", "value", "isNew", "onClick", "selected", "title"]);
  var _useContext = (0, _react.useContext)(_context.default),
    prefix = _useContext.prefix;
  var clsPrefix = prefix + "select-option";
  return /*#__PURE__*/(0, _react.createElement)(_list.default.Item, _extends({}, others, {
    className: (0, _classnames.default)(clsPrefix, className, (_classNames = {}, _classNames[clsPrefix + "--disabled"] = disabled, _classNames)),
    title: children || title,
    media: /*#__PURE__*/(0, _react.createElement)(_list.default.ItemMedia, null, /*#__PURE__*/(0, _react.createElement)(_checkbox.default, {
      checked: selected,
      disabled: disabled || isNew,
      className: clsPrefix + "-checkbox"
    })),
    onClick: function onClick() {
      if (!disabled) {
        _onClick(value);
      }
    }
  }));
};
Option.displayName = 'SelectOption';
var _default = exports.default = /*#__PURE__*/(0, _react.forwardRef)(Option);
