// @alilc/lowcode-designer/lib/builtin-simulator/bem-tools/border-container.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.BorderContainerInstance = exports.BorderContainer = void 0;
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _classnames = _interopRequireDefault(require("classnames"));
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _lowcodeUtils = require("@alilc/lowcode-utils");
var _class, _class2;
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
var BorderContainerInstance = exports.BorderContainerInstance = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(BorderContainerInstance, _PureComponent);
  function BorderContainerInstance() {
    return _PureComponent.apply(this, arguments) || this;
  }
  var _proto = BorderContainerInstance.prototype;
  _proto.render = function render() {
    var _this$props = this.props,
      title = _this$props.title,
      rect = _this$props.rect,
      scale = _this$props.scale,
      scrollX = _this$props.scrollX,
      scrollY = _this$props.scrollY;
    if (!rect) {
      return null;
    }
    var style = {
      width: rect.width * scale,
      height: rect.height * scale,
      transform: "translate(" + (scrollX + rect.left) * scale + "px, " + (scrollY + rect.top) * scale + "px)"
    };
    var className = (0, _classnames["default"])('lc-borders lc-borders-detecting');
    return /*#__PURE__*/React.createElement("div", {
      className: className,
      style: style
    }, /*#__PURE__*/React.createElement(_lowcodeEditorCore.Title, {
      title: title,
      className: "lc-borders-title"
    }));
  };
  return BorderContainerInstance;
}(_react.PureComponent);
function getTitle(title) {
  if (typeof title === 'string') return title;
  if ((0, _lowcodeUtils.isI18nData)(title)) {
    var locale = _lowcodeEditorCore.globalLocale.getLocale() || 'zh-CN';
    return "\u5C06\u653E\u5165\u5230\u6B64" + title[locale];
  }
  return '';
}
var BorderContainer = exports.BorderContainer = (0, _lowcodeEditorCore.observer)(_class = (_class2 = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(BorderContainer, _Component);
  function BorderContainer() {
    var _this;
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {};
    return _this;
  }
  var _proto2 = BorderContainer.prototype;
  _proto2.componentDidMount = function componentDidMount() {
    var _this2 = this;
    var host = this.props.host;
    host.designer.editor.eventBus.on('designer.dropLocation.change', function (loc) {
      var target = _this2.state.target;
      if (target === (loc === null || loc === void 0 ? void 0 : loc.target)) return;
      _this2.setState({
        target: loc === null || loc === void 0 ? void 0 : loc.target
      });
    });
  };
  _proto2.render = function render() {
    var _this3 = this;
    var host = this.props.host;
    var target = this.state.target;
    if (target == undefined) {
      return null;
    }
    var instances = host.getComponentInstances(target);
    if (!instances || instances.length < 1) {
      return null;
    }
    if (instances.length === 1) {
      return /*#__PURE__*/React.createElement(BorderContainerInstance, {
        key: "line-h",
        title: getTitle(target.componentMeta.title),
        scale: this.scale,
        scrollX: this.scrollX,
        scrollY: this.scrollY,
        rect: host.computeComponentInstanceRect(instances[0], target.componentMeta.rootSelector)
      });
    }
    return /*#__PURE__*/React.createElement(_react.Fragment, null, instances.map(function (inst, i) {
      return /*#__PURE__*/React.createElement(BorderContainerInstance, {
        key: "line-h-" + i,
        title: getTitle(target.componentMeta.title),
        scale: _this3.scale,
        scrollX: _this3.scrollX,
        scrollY: _this3.scrollY,
        rect: host.computeComponentInstanceRect(inst, target.componentMeta.rootSelector)
      });
    }));
  };
  (0, _createClass2["default"])(BorderContainer, [{
    key: "scale",
    get: function get() {
      return this.props.host.viewport.scale;
    }
  }, {
    key: "scrollX",
    get: function get() {
      return this.props.host.viewport.scrollX;
    }
  }, {
    key: "scrollY",
    get: function get() {
      return this.props.host.viewport.scrollY;
    }
  }]);
  return BorderContainer;
}(_react.Component), ((0, _applyDecoratedDescriptor2["default"])(_class2.prototype, "scale", [_lowcodeEditorCore.computed], Object.getOwnPropertyDescriptor(_class2.prototype, "scale"), _class2.prototype), (0, _applyDecoratedDescriptor2["default"])(_class2.prototype, "scrollX", [_lowcodeEditorCore.computed], Object.getOwnPropertyDescriptor(_class2.prototype, "scrollX"), _class2.prototype), (0, _applyDecoratedDescriptor2["default"])(_class2.prototype, "scrollY", [_lowcodeEditorCore.computed], Object.getOwnPropertyDescriptor(_class2.prototype, "scrollY"), _class2.prototype)), _class2)) || _class;
