// @alilc/lowcode-designer/lib/builtin-simulator/bem-tools/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.BemTools = void 0;
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = _interopRequireWildcard(require("react"));
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _borderDetecting = require("./border-detecting");
var _borderContainer = require("./border-container");
var _borderSelecting = require("./border-selecting");
var _borderResizing = _interopRequireDefault(require("./border-resizing"));
var _insertion = require("./insertion");
require("./bem-tools.less");
require("./borders.less");
var _class;
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
var BemTools = exports.BemTools = (0, _lowcodeEditorCore.observer)(_class = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(BemTools, _Component);
  function BemTools() {
    return _Component.apply(this, arguments) || this;
  }
  var _proto = BemTools.prototype;
  _proto.render = function render() {
    var host = this.props.host;
    var designMode = host.designMode;
    var _host$viewport = host.viewport,
      scrollX = _host$viewport.scrollX,
      scrollY = _host$viewport.scrollY,
      scale = _host$viewport.scale;
    if (designMode === 'live') {
      return null;
    }
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: "lc-bem-tools",
      style: {
        transform: "translate(" + -scrollX * scale + "px," + -scrollY * scale + "px)"
      }
    }, !_lowcodeEditorCore.engineConfig.get('disableDetecting') && /*#__PURE__*/_react["default"].createElement(_borderDetecting.BorderDetecting, {
      key: "hovering",
      host: host
    }), /*#__PURE__*/_react["default"].createElement(_borderSelecting.BorderSelecting, {
      key: "selecting",
      host: host
    }), _lowcodeEditorCore.engineConfig.get('enableReactiveContainer') && /*#__PURE__*/_react["default"].createElement(_borderContainer.BorderContainer, {
      key: "reactive-container-border",
      host: host
    }), /*#__PURE__*/_react["default"].createElement(_insertion.InsertionView, {
      key: "insertion",
      host: host
    }), /*#__PURE__*/_react["default"].createElement(_borderResizing["default"], {
      key: "resizing",
      host: host
    }), host.designer.bemToolsManager.getAllBemTools().map(function (tools) {
      var ToolsCls = tools.item;
      return /*#__PURE__*/_react["default"].createElement(ToolsCls, {
        key: tools.name,
        host: host
      });
    }));
  };
  return BemTools;
}(_react.Component)) || _class;
