// @alilc/lowcode-designer/lib/builtin-simulator/bem-tools/manager.js
"use strict";

exports.__esModule = true;
exports.BemToolsManager = void 0;
var _utils = require("../../utils");
var BemToolsManager = exports.BemToolsManager = /*#__PURE__*/function () {
  function BemToolsManager(designer) {
    this.designer = void 0;
    this.toolsContainer = [];
    this.designer = designer;
  }
  var _proto = BemToolsManager.prototype;
  _proto.addBemTools = function addBemTools(toolsData) {
    var found = this.toolsContainer.find(function (item) {
      return item.name === toolsData.name;
    });
    (0, _utils.invariant)(!found, toolsData.name + " already exists");
    this.toolsContainer.push(toolsData);
  };
  _proto.removeBemTools = function removeBemTools(name) {
    var index = this.toolsContainer.findIndex(function (item) {
      return item.name === name;
    });
    if (index !== -1) {
      this.toolsContainer.splice(index, 1);
    }
  };
  _proto.getAllBemTools = function getAllBemTools() {
    return this.toolsContainer;
  };
  return BemToolsManager;
}();
