// universal-request/es/miniapp-bytedance/index.js
import { ERROR_REQUEST_TIMEOUT, ERROR_REQUEST_ABORT } from '../types';
import { object2json } from '../utils';
export default function (options) {
  return new Promise(function (resolve, reject) {
    var url = options.url,
        method = options.method,
        data = options.data,
        dataType = options.dataType,
        headers = options.headers,
        timeout = options.timeout;
    var timeoutTimer;
    var requestTask = tt.request({
      url: url,
      headers: headers,
      method: method,
      data: data,
      timeout: timeout,
      dataType: dataType,
      success: function success(res) {
        resolve(res);
      },
      fail: function fail(res) {
        reject({
          code: ERROR_REQUEST_ABORT.code,
          message: object2json(res)
        });
      },
      complete: function complete() {
        if (timeoutTimer) {
          clearTimeout(timeoutTimer);
        }
      }
    });

    if (timeout) {
      timeoutTimer = setTimeout(function () {
        requestTask.abort();
        reject(ERROR_REQUEST_TIMEOUT);
      }, timeout);
    }
  });
}
