// @alilc/lowcode-editor-core/lib/widgets/tip/help-tips.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.HelpTip = HelpTip;
var _icon = _interopRequireDefault(require("@alifd/next/lib/icon"));
var _tip = require("./tip");
function HelpTip(_ref) {
  var help = _ref.help,
    _ref$direction = _ref.direction,
    direction = _ref$direction === void 0 ? 'top' : _ref$direction,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'small' : _ref$size;
  if (typeof help === 'string') {
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_icon["default"], {
      type: "help",
      size: size,
      className: "lc-help-tip"
    }), /*#__PURE__*/React.createElement(_tip.Tip, {
      direction: direction
    }, help));
  }
  if (typeof help === 'object' && help.url) {
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("a", {
      href: help.url,
      target: "_blank",
      rel: "noopener noreferrer"
    }, /*#__PURE__*/React.createElement(_icon["default"], {
      type: "help",
      size: size,
      className: "lc-help-tip"
    })), /*#__PURE__*/React.createElement(_tip.Tip, {
      direction: direction
    }, help.content));
  }
  return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_icon["default"], {
    type: "help",
    size: "small",
    className: "lc-help-tip"
  }), /*#__PURE__*/React.createElement(_tip.Tip, {
    direction: direction
  }, help.content));
}
