// @alilc/lowcode-plugin-code-generator/lib/components/CodeGenPreview/CodeGenPreview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.CodeGenPreview = CodeGenPreview;

var _react = _interopRequireDefault(require("react"));

var _next = require("@alifd/next");

var _useIsInIntranet = require("./useIsInIntranet");

var _GravityPreview = require("../GravityPreview");

var _CodeSandboxPreview = require("../CodeSandboxPreview");

var _ForceUpdate = require("../GravityPreview/ForceUpdate");

require("./CodeGenPreview.scss");

function CodeGenPreview(props) {
  var isInIntranet = (0, _useIsInIntranet.useIsInIntranet)();
  var previewPanes = [{
    title: '基于 Gravity 预览',
    available: isInIntranet,
    render: function render() {
      return /*#__PURE__*/_react["default"].createElement(_GravityPreview.GravityPreview, props);
    }
  }, {
    title: '基于 CodeSandbox 预览',
    available: true,
    render: function render() {
      return /*#__PURE__*/_react["default"].createElement(_CodeSandboxPreview.CodeSandboxPreview, props);
    }
  }];
  var availablePreviewPanes = previewPanes.filter(function (pane) {
    return pane.available;
  });

  if (availablePreviewPanes.length === 1) {
    return availablePreviewPanes[0].render();
  }

  return /*#__PURE__*/_react["default"].createElement(_next.Tab, {
    className: "code-gen-preview-tabs",
    size: "small"
  }, availablePreviewPanes.map(function (pane) {
    return /*#__PURE__*/_react["default"].createElement(_next.Tab.Item, {
      title: pane.title,
      key: pane.title
    }, /*#__PURE__*/_react["default"].createElement(_ForceUpdate.ForceUpdate, {
      watchKey: props.refresh
    }, pane.render()));
  }));
}
