// @alilc/lowcode-plugin-code-generator/lib/components/CodeGenPreview/fixPreviewCode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.fixPreviewCode = fixPreviewCode;

var _extends4 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

// 默认生成的代码不符合在线预览的规则，需要处理下
function fixPreviewCode(code, unused) {
  var _JSON$parse, _fixedCode$modules$P;

  if (!code) {
    return null;
  }

  var fixedModules = Object.entries(code.modules || {}) // 去掉隐藏文件和 html 以及 md 等无用文件:
  .filter(function (_ref) {
    var fpath = _ref[0];
    return !fpath.startsWith('/.') && !fpath.endsWith('.html') && !fpath.endsWith('.md');
  }) // 所有的 jsx 文件改成 js 文件（因为 gravity 不支持 jsx）
  .map(function (_ref2) {
    var fpath = _ref2[0],
        module = _ref2[1];
    return [fpath.replace(/\.jsx$/, '.js'), module];
  }).map(function (_ref3) {
    var fpath = _ref3[0],
        module = _ref3[1];
    return [fpath.replace(/\.scss$/, '.css'), module];
  }).reduce(function (acc, _ref4) {
    var _extends2;

    var fpath = _ref4[0],
        module = _ref4[1];
    return (0, _extends4["default"])({}, acc, (_extends2 = {}, _extends2[fpath] = (0, _extends4["default"])({}, module, {
      fpath: fpath,
      code: module.code // 所有的 import xxx from '@/yyy' 转换为 import xxx from '/src/yyy'
      .replace(/import\s+([^\s]+)\s+from\s+['"]@\/([^'"]+)['"]/g, function (_, name, path) {
        return "import " + name + " from '/src/" + path + "'";
      }) // 所有的 import styles from 'xxx.scss' 转换为 import styles from '/src/xxx.css'
      .replace(/import\s+([^\s]+)\s+from\s+['"]([^'"]+)\.scss['"]/g, function (_, name, path) {
        return "import " + name + " from '" + path + ".css'";
      })
    }), _extends2));
  }, {});
  var fixedCode = (0, _extends4["default"])({}, code, {
    modules: fixedModules
  });
  fixedCode.modules['/src/app.js'] = (0, _extends4["default"])({}, fixedCode.modules['/src/app.js'], {
    code: "\n// \u6CE8\u610F\uFF1A\u4E3A\u4E86\u65B9\u4FBF\u5728\u6D4F\u89C8\u5668\u4E2D\u8FDB\u884C\u9884\u89C8\uFF0C\u8FD9\u91CC\u7B80\u5316\u4E86\u4E00\u4E9B\u5185\u5BB9\nimport './shims';\nimport './global.css';\n\nimport React from 'react';\nimport ReactDOM from 'react-dom';\n\nimport Page from '" + Object.keys(fixedModules).find(function (fpath) {
      return fpath.startsWith('/src/pages/');
    }) + "';\n\nReactDOM.render(<Page/>, document.getElementById('root'));\n"
  }); // 一些垫片，用于解决一些不能被 gravity 解析的问题
  // 1. @alilc/lowcode-materials 中有部分代码直接使用了全局变量 PropTypes...

  fixedCode.modules['/src/shims.js'] = {
    fpath: '/src/shims.js',
    code: "\n// \u4E00\u4E9B\u57AB\u7247\u4EE3\u7801\uFF0C\u7528\u6765\u89E3\u51B3\u76F4\u63A5\u5728\u6D4F\u89C8\u5668\u4E2D\u9884\u89C8\u7684\u65F6\u5019\u7684\u95EE\u9898\nimport PropTypes from 'prop-types';\n\nwindow.PropTypes = PropTypes;\n"
  };
  fixedCode.modules['/package.json'] = (0, _extends4["default"])({}, fixedCode.modules['/package.json'], {
    code: JSON.stringify({
      name: 'demo',
      version: '1.0.0',
      dependencies: (0, _extends4["default"])({
        react: '^16.8.3',
        'react-dom': '^16.8.3'
      }, (_JSON$parse = JSON.parse(((_fixedCode$modules$P = fixedCode.modules['/package.json']) === null || _fixedCode$modules$P === void 0 ? void 0 : _fixedCode$modules$P.code) || '{}')) === null || _JSON$parse === void 0 ? void 0 : _JSON$parse.dependencies)
    })
  });
  fixedCode.modules = pickKeys(fixedCode.modules, [// '/tsconfig.json',
  // '/jsconfig.json',
  // '/build.json',
  // '/abc.json',
  '/package.json', '/src/routes.js', '/src/app.js', '/src/constants.js', '/src/utils.js', '/src/i18n.js', '/src/global.css', '/src/index.js', '/src/shims.js'].concat(Object.keys(fixedModules).filter(function (fpath) {
    return fpath.startsWith('/src/pages/');
  })));
  fixedCode.modules['/src/routes.js'] = (0, _extends4["default"])({}, fixedCode.modules['/src/routes.js'], {
    code: "\n"
  });
  fixedCode.modules['/src/global.css'] = {
    fpath: '/src/global.css',
    code: "\nbody {\n  -webkit-font-smoothing: antialiased;\n}\n"
  };
  Object.assign(fixedCode.modules, {
    '/src/index.html': {
      code: '<div id="root"></div>',
      fpath: '/src/index.html'
    },
    '/src/index.less': {
      fpath: '/src/index.less',
      code: "\n@import \"~@alifd/next/dist/next.css\";\n@import '~@alifd/pro-layout/dist/AlifdProLayout.css';\n"
    }
  });
  fixedCode.type = 'riddle';
  return fixedCode;
}

function pickKeys(data, keys) {
  return keys.reduce(function (acc, key) {
    if (data[key]) {
      var _extends3;

      return (0, _extends4["default"])({}, acc, (_extends3 = {}, _extends3[key] = data[key], _extends3));
    }

    return acc;
  }, {});
}
