// @alilc/lowcode-plugin-code-editor/lib/components/CssEditor/CssEditor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.CssEditor = void 0;
var _tab = _interopRequireDefault(require("@alifd/next/lib/tab"));
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = _interopRequireWildcard(require("react"));
var _lowcodePluginBaseMonacoEditor = _interopRequireDefault(require("@alilc/lowcode-plugin-base-monaco-editor"));
var _config = require("../../config");
require("./CssEditor.less");
var _utils = require("../../utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
// TODO: CSS语法提示后续再添加
var CssEditor = exports.CssEditor = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(CssEditor, _PureComponent);
  function CssEditor() {
    var _this$props$cssCode;
    var _this;
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.state = {
      code: (_this$props$cssCode = _this.props.cssCode) !== null && _this$props$cssCode !== void 0 ? _this$props$cssCode : ''
    };
    _this.cssEditor = void 0;
    return _this;
  }
  var _proto = CssEditor.prototype;
  _proto.editorDidMount = function editorDidMount(editor) {
    this.cssEditor = editor;
  };
  _proto.getBeautifiedCSS = function getBeautifiedCSS() {
    var code = this.state.code;
    var nextCode = (0, _utils.beautifyCSS)(code);
    // css 将会经过格式化，因此需要同步回来
    this.setState({
      code: nextCode
    });
    return nextCode;
  };
  _proto.render = function render() {
    var _this2 = this;
    var code = this.state.code;
    return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(_tab["default"], {
      size: "small",
      shape: "wrapped",
      activeKey: this.props.currentTab
    }, /*#__PURE__*/_react["default"].createElement(_tab["default"].Item, {
      title: (code !== this.props.cssCode ? '* ' : '') + "index.css",
      key: _config.TAB_KEY.CSS,
      onClick: function onClick() {
        return _this2.props.onTabChange(_config.TAB_KEY.CSS);
      }
    })), this.props.currentTab === _config.TAB_KEY.CSS && /*#__PURE__*/_react["default"].createElement("div", {
      className: "plugin-code-editor-css plugin-code-editor-inner"
    }, /*#__PURE__*/_react["default"].createElement(_lowcodePluginBaseMonacoEditor["default"], {
      value: code,
      language: "css",
      height: "100%",
      supportFullScreen: true,
      onChange: function onChange(newCode) {
        _this2._updateCode(newCode);
      },
      editorDidMount: function editorDidMount(useMonaco, editor) {
        _this2.editorDidMount(editor);
      }
    })));
  };
  _proto._updateCode = function _updateCode(newCode) {
    this.setState({
      code: newCode
    });
  };
  return CssEditor;
}(_react.PureComponent);
CssEditor.defaultProps = void 0;
