// @alilc/lowcode-plugin-set-ref-prop/es/index.js
import _extends from "@babel/runtime/helpers/extends";
import { v4 as uuidv4 } from 'uuid';
import { material } from '@alilc/lowcode-engine';
function addonCombine(metadata) {
  var _configure$combined;
  var componentName = metadata.componentName,
    _metadata$configure = metadata.configure,
    configure = _metadata$configure === void 0 ? {} : _metadata$configure;
  var isRoot = componentName === 'Page' || componentName === 'Component';
  if (isRoot) {
    return metadata;
  }
  var advancedGroup;
  var refItem = {
    title: {
      label: 'refId',
      tip: '用于获取组件实例，调用物料内部方法',
      icon: ''
    },
    name: 'ref',
    setter: {
      componentName: 'StringSetter'
    },
    defaultValue: function defaultValue() {
      var uuid = uuidv4().replace('-', '').substring(0, 8);
      return componentName.toLowerCase() + "-" + uuid;
    },
    extraProps: {
      display: 'block',
      supportVariable: false
    }
  };
  if (!configure.combined) {
    configure.combined = [];
  }
  advancedGroup = (_configure$combined = configure.combined) === null || _configure$combined === void 0 ? void 0 : _configure$combined.filter(function (d) {
    return d.name === '#advanced';
  })[0];
  if (!advancedGroup) {
    advancedGroup = {
      name: '#advanced',
      title: {
        type: 'i18n',
        'zh-CN': '高级',
        'en-US': 'Advanced'
      },
      items: [refItem]
    };
    configure.combined.push(advancedGroup);
  }
  if (!advancedGroup.items) {
    advancedGroup.items = [refItem];
  }
  var advanceItems = advancedGroup.items || [];
  if (!advanceItems || !advanceItems.length || !(advanceItems !== null && advanceItems !== void 0 && advanceItems.filter(function (d) {
    return d.name === 'ref';
  }).length)) {
    advanceItems.push(refItem);
  }
  return _extends({}, metadata, {
    configure: configure
  });
}
var SetRefPropPlugin = function SetRefPropPlugin() {
  return {
    init: function init() {
      material.registerMetadataTransducer(addonCombine, 110, 'register-ref-prop');
    }
  };
};
SetRefPropPlugin.pluginName = 'SetRefPropPlugin';
export default SetRefPropPlugin;
