// @alilc/lowcode-engine-ext/lib/setter/style-setter/components/color-input/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports["default"] = void 0;
var _balloon = _interopRequireDefault(require("@alifd/next/lib/balloon"));
var _input = _interopRequireDefault(require("@alifd/next/lib/input"));
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var React = _interopRequireWildcard(require("react"));
var _reactColor = require("react-color");
require("./index.less");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var ColorSetter = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2["default"])(ColorSetter, _React$Component);
  function ColorSetter(props) {
    var _this;
    _this = _React$Component.call(this, props) || this;
    /**
     * 屏幕分辨率监听
     */
    _this.screenChange = function () {
      window.addEventListener('resize', _this.changeWidth);
    };
    /**
     * 屏幕分辨率 变换 =>  改变冒泡框的位置
     */
    _this.changeWidth = function () {
      _this.setState({
        width: document.body.clientWidth < 1860 ? -92 : -138
      });
    };
    _this.inputChange = function (color) {
      var _this$props = _this.props,
        onStyleChange = _this$props.onStyleChange,
        styleKey = _this$props.styleKey;
      if (color == '') {
        onStyleChange([{
          styleKey: styleKey,
          value: null
        }]);
      }
    };
    _this.handleChange = function (color) {
      var _this$props2 = _this.props,
        onStyleChange = _this$props2.onStyleChange,
        styleKey = _this$props2.styleKey;
      var rgb = color.rgb,
        hex = color.hex;
      var r = rgb.r,
        g = rgb.g,
        b = rgb.b,
        a = rgb.a;
      if (a === 1) {
        onStyleChange([{
          styleKey: styleKey,
          value: hex
        }]);
      } else {
        onStyleChange([{
          styleKey: styleKey,
          value: "rgba(" + r + "," + g + "," + b + "," + a + ")"
        }]);
      }
    };
    _this.state = {
      width: 50
    };
    return _this;
  }
  var _proto = ColorSetter.prototype;
  _proto.componentDidMount = function componentDidMount() {
    this.screenChange();
    this.changeWidth();
    // const { onChange, value, defaultValue } = this.props;
    // if (value == undefined && defaultValue) {
    //   onChange(defaultValue);
    // }
  };
  _proto.componentWillUnmount = function componentWillUnmount() {
    window.removeEventListener('resize', this.changeWidth);
  };
  _proto.render = function render() {
    var _this$props3 = this.props,
      styleKey = _this$props3.styleKey,
      styleData = _this$props3.styleData,
      _this$props3$inputWid = _this$props3.inputWidth,
      inputWidth = _this$props3$inputWid === void 0 ? '108px' : _this$props3$inputWid,
      color = _this$props3.color;
    var InputTarget = /*#__PURE__*/React.createElement(_input["default"], {
      className: "lowcode-setter-color",
      style: {
        width: inputWidth
      },
      hasClear: true,
      innerBefore: /*#__PURE__*/React.createElement("div", {
        className: "color-box",
        style: {
          backgroundColor: color ? color : styleData[styleKey]
        }
      }),
      onChange: this.inputChange,
      value: color ? color : styleData[styleKey]
    });
    return /*#__PURE__*/React.createElement(_balloon["default"], {
      needAdjust: true,
      align: "tr",
      offset: [-30, 0],
      style: {
        padding: 0
      },
      trigger: InputTarget,
      triggerType: "click",
      closable: false
    }, /*#__PURE__*/React.createElement(_reactColor.SketchPicker, {
      width: 250,
      color: color ? color : styleData[styleKey],
      onChange: this.handleChange
    }));
  };
  return ColorSetter;
}(React.Component);
exports["default"] = ColorSetter;
