// @alilc/lowcode-engine-ext/lib/setter/style-setter/pro/layout/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports["default"] = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _collapse = _interopRequireDefault(require("@alifd/next/lib/collapse"));
var React = _interopRequireWildcard(require("react"));
var _row = _interopRequireDefault(require("../../components/row"));
var _layoutBox = _interopRequireDefault(require("./layoutBox"));
var _number = _interopRequireDefault(require("../../components/number"));
var _config = _interopRequireDefault(require("./config.json"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var Panel = _collapse["default"].Panel;
var defaultLayoutPropsConfig = {
  // display 展示列表
  showDisPlayList: ['inline', 'flex', 'block', 'inline-block', 'none'],
  isShowPadding: true,
  isShowMargin: true,
  isShowWidthHeight: true
};
var _default = function _default(props) {
  var onStyleChange = props.onStyleChange,
    styleData = props.styleData,
    layoutPropsConfig = props.layoutPropsConfig,
    unit = props.unit;

  // 配置合并
  var propsConfig = (0, _extends2["default"])({}, defaultLayoutPropsConfig, layoutPropsConfig);

  // 传入配置
  var showDisPlayList = propsConfig.showDisPlayList,
    isShowWidthHeight = propsConfig.isShowWidthHeight;
  // 静态配置
  var display = _config["default"].display,
    flexDirection = _config["default"].flexDirection,
    justifyContent = _config["default"].justifyContent,
    alignItems = _config["default"].alignItems,
    flexWrap = _config["default"].flexWrap;
  var displayDataList = display.dataList.filter(function (item) {
    return showDisPlayList.indexOf(item.value) >= 0;
  });
  return /*#__PURE__*/React.createElement(_collapse["default"], {
    defaultExpandedKeys: ['0']
  }, /*#__PURE__*/React.createElement(Panel, {
    title: "\u5E03\u5C40",
    className: "layout-style-container"
  }, /*#__PURE__*/React.createElement(_row["default"], (0, _extends2["default"])({
    title: display.title,
    dataList: displayDataList,
    styleKey: "display"
  }, props, {
    longTitle: true
  })), styleData['display'] === 'flex' && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(_row["default"], (0, _extends2["default"])({
    title: flexDirection.title,
    dataList: flexDirection.dataList,
    styleKey: "flexDirection",
    longTitle: true
  }, props)), /*#__PURE__*/React.createElement(_row["default"], (0, _extends2["default"])({
    title: justifyContent.title,
    dataList: justifyContent.dataList,
    styleKey: "justifyContent",
    longTitle: true
  }, props)), /*#__PURE__*/React.createElement(_row["default"], (0, _extends2["default"])({
    title: alignItems.title,
    dataList: alignItems.dataList,
    styleKey: "alignItems",
    longTitle: true
  }, props)), /*#__PURE__*/React.createElement(_row["default"], (0, _extends2["default"])({
    title: flexWrap.title,
    dataList: flexWrap.dataList,
    styleKey: "flexWrap",
    longTitle: true
  }, props))), /*#__PURE__*/React.createElement(_layoutBox["default"], {
    styleData: styleData,
    onStyleChange: onStyleChange,
    layoutPropsConfig: propsConfig,
    unit: unit
  }), isShowWidthHeight && /*#__PURE__*/React.createElement("div", {
    className: "inner-row-contaienr"
  }, /*#__PURE__*/React.createElement("div", {
    className: "row-item"
  }, /*#__PURE__*/React.createElement("span", {
    className: "row-item-title"
  }, "\u5BBD\u5EA6"), /*#__PURE__*/React.createElement(_number["default"], (0, _extends2["default"])({
    style: {
      marginRight: '10px',
      width: '100%'
    },
    min: 0,
    styleKey: "width"
  }, props, {
    unit: ['px', '%'],
    useComputedStyle: true
  }))), /*#__PURE__*/React.createElement("div", {
    className: "row-item"
  }, /*#__PURE__*/React.createElement("span", {
    className: "row-item-title"
  }, "\u9AD8\u5EA6"), /*#__PURE__*/React.createElement(_number["default"], (0, _extends2["default"])({
    styleKey: "height",
    min: 0
  }, props, {
    style: {
      width: '100%'
    },
    unit: ['px', '%'],
    useComputedStyle: true
  }))))));
};
exports["default"] = _default;
