// @alifd/meet-react/lib/card/card.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
var _babelRuntimeJsxStyleTransform = require("@alifd/babel-runtime-jsx-style-transform");
var _tslib = require("tslib");
var _react = _interopRequireWildcard(require("react"));
var _classnames = _interopRequireDefault(require("classnames"));
var _view = _interopRequireDefault(require("../view"));
var _icon = _interopRequireDefault(require("../icon"));
var _utils = require("../utils");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
var ARROW_ICON_MAP = {
  right: 'arrow-right',
  up: 'arrow-up',
  down: 'arrow-down',
  loading: 'loading'
};
var Card = function (props) {
  var _classNames;
  var _props$prefix = props.prefix,
    prefix = _props$prefix === void 0 ? 'mt-' : _props$prefix,
    _props$showHeadDivide = props.showHeadDivider,
    showHeadDivider = _props$showHeadDivide === void 0 ? true : _props$showHeadDivide,
    media = props.media,
    arrow = props.arrow,
    extra = props.extra,
    title = props.title,
    className = props.className,
    headerClassName = props.headerClassName,
    headerStyle = props.headerStyle,
    bodyClassName = props.bodyClassName,
    footerClassName = props.footerClassName,
    contentHeight = props.contentHeight,
    _props$bodyStyle = props.bodyStyle,
    bodyStyle = _props$bodyStyle === void 0 ? {} : _props$bodyStyle,
    _props$footerStyle = props.footerStyle,
    footerStyle = _props$footerStyle === void 0 ? {} : _props$footerStyle,
    actions = props.actions,
    children = props.children,
    _props$onHeaderClick = props.onHeaderClick,
    onHeaderClick = _props$onHeaderClick === void 0 ? _utils.noop : _props$onHeaderClick,
    _props$onBodyClick = props.onBodyClick,
    onBodyClick = _props$onBodyClick === void 0 ? _utils.noop : _props$onBodyClick,
    others = (0, _tslib.__rest)(props, ["prefix", "showHeadDivider", "media", "arrow", "extra", "title", "className", "headerClassName", "headerStyle", "bodyClassName", "footerClassName", "contentHeight", "bodyStyle", "footerStyle", "actions", "children", "onHeaderClick", "onBodyClick"]);
  var clsPrefix = prefix + "card";
  var clses = {
    card: (0, _classnames.default)(className, clsPrefix),
    header: (0, _classnames.default)(clsPrefix + "-header", headerClassName, (_classNames = {}, _classNames[clsPrefix + "-header--no-divider"] = !showHeadDivider, _classNames)),
    title: clsPrefix + "-title",
    media: clsPrefix + "-media",
    extra: clsPrefix + "-extra",
    arrow: clsPrefix + "-arrow",
    body: (0, _classnames.default)(clsPrefix + "-body", bodyClassName),
    footer: (0, _classnames.default)(clsPrefix + "-footer", footerClassName)
  };
  var _bodyStyle = bodyStyle;
  if ((0, _utils.isNumber)(contentHeight) && contentHeight >= 0) {
    _bodyStyle.height = contentHeight;
  }
  return /*#__PURE__*/(0, _react.createElement)(_view.default, _extends({
    className: clses.card
  }, others), !(0, _utils.isInvalidValue)(title) && !!title ? /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: clses.header,
    onClick: onHeaderClick,
    style: (0, _babelRuntimeJsxStyleTransform.rpx2vw4style)(headerStyle)
  }, media ? /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: clses.media
  }, media) : null, /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: clses.title
  }, title), extra ? /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: clses.extra
  }, extra) : null, Object.keys(ARROW_ICON_MAP).indexOf(arrow) > -1 ? /*#__PURE__*/(0, _react.createElement)(_icon.default, {
    className: clses.arrow,
    type: ARROW_ICON_MAP[arrow],
    size: "s"
  }) : null) : null, children ? /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: clses.body,
    style: (0, _babelRuntimeJsxStyleTransform.rpx2vw4style)(_bodyStyle),
    onClick: onBodyClick
  }, children) : null, (0, _utils.isValidArray)(actions) ? /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: clses.footer,
    style: (0, _babelRuntimeJsxStyleTransform.rpx2vw4style)(footerStyle)
  }, actions) : null);
};
Card.displayName = 'Card';
var _default = exports.default = Card;
