// @alilc/lowcode-plugin-code-generator/es/components/CodeGenPreview/CodeGenPreview.js
import React from 'react';
import { Tab } from '@alifd/next';
import { useIsInIntranet } from './useIsInIntranet';
import { GravityPreview } from '../GravityPreview';
import { CodeSandboxPreview } from '../CodeSandboxPreview';
import { ForceUpdate } from '../GravityPreview/ForceUpdate';
import './CodeGenPreview.scss';
export function CodeGenPreview(props) {
  var isInIntranet = useIsInIntranet();
  var previewPanes = [{
    title: '基于 Gravity 预览',
    available: isInIntranet,
    render: function render() {
      return /*#__PURE__*/React.createElement(GravityPreview, props);
    }
  }, {
    title: '基于 CodeSandbox 预览',
    available: true,
    render: function render() {
      return /*#__PURE__*/React.createElement(CodeSandboxPreview, props);
    }
  }];
  var availablePreviewPanes = previewPanes.filter(function (pane) {
    return pane.available;
  });

  if (availablePreviewPanes.length === 1) {
    return availablePreviewPanes[0].render();
  }

  return /*#__PURE__*/React.createElement(Tab, {
    className: "code-gen-preview-tabs",
    size: "small"
  }, availablePreviewPanes.map(function (pane) {
    return /*#__PURE__*/React.createElement(Tab.Item, {
      title: pane.title,
      key: pane.title
    }, /*#__PURE__*/React.createElement(ForceUpdate, {
      watchKey: props.refresh
    }, pane.render()));
  }));
}
