// @alilc/lowcode-plugin-code-generator/lib/components/CodeGenResult/CodeGenResult.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.CodeGenResult = CodeGenResult;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _extends3 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

require("./CodeGenResult.scss");

var _react = _interopRequireWildcard(require("react"));

var _fileSaver = _interopRequireDefault(require("file-saver"));

var _jszip = _interopRequireDefault(require("jszip"));

var _next = require("@alifd/next");

var _CodeGenPreview = require("../CodeGenPreview");

var _SourcesView = require("../SourcesView");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function CodeGenResult(_ref) {
  var result = _ref.result,
      schema = _ref.schema;

  var _useState = (0, _react.useState)({
    expandedKeys: ['sources', 'preview']
  }),
      paneState = _useState[0],
      setPaneState = _useState[1];

  var _useState2 = (0, _react.useState)(null),
      gravityCode = _useState2[0],
      setGravityCode = _useState2[1];

  var _useState3 = (0, _react.useState)(0),
      refresh = _useState3[0],
      setRefresh = _useState3[1];

  (0, _react.useEffect)(function () {
    setGravityCode(convertCodeGenResult(result, schema));
  }, [result]);

  if (!result) {
    return null;
  }

  var sourcesViewHeight = paneState.expandedKeys.includes('preview') ? '40vh' : '80vh';
  var gravityDemoHeight = paneState.expandedKeys.includes('sources') ? '40vh' : '80vh';

  var handleDownloadSources = /*#__PURE__*/function () {
    var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(e) {
      var zip;
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.prev = 0;
              e.preventDefault();
              e.stopPropagation();
              zip = new _jszip["default"]();
              Object.values((gravityCode === null || gravityCode === void 0 ? void 0 : gravityCode.modules) || {}).forEach(function (file) {
                zip.file(file.fpath.replace(/^\/+/, ''), file.code);
              });
              _context.next = 7;
              return zip.generateAsync({
                type: 'blob'
              }).then(function (content) {
                _fileSaver["default"].saveAs(content, 'ali-lowcode-generated-sources.zip');
              });

            case 7:
              _context.next = 13;
              break;

            case 9:
              _context.prev = 9;
              _context.t0 = _context["catch"](0);
              console.log('failed to download sources: ', _context.t0);

              _next.Message.error('下载失败！');

            case 13:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, null, [[0, 9]]);
    }));

    return function handleDownloadSources(_x) {
      return _ref2.apply(this, arguments);
    };
  }();

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: "code-gen-result"
  }, /*#__PURE__*/_react["default"].createElement(_next.Collapse, {
    expandedKeys: paneState.expandedKeys,
    onExpand: function onExpand(expandedKeys) {
      setPaneState({
        expandedKeys: expandedKeys
      });
    }
  }, /*#__PURE__*/_react["default"].createElement(_next.Collapse.Panel, {
    title: /*#__PURE__*/_react["default"].createElement("span", null, "\u51FA\u7801\u751F\u6210\u7684\u6E90\u4EE3\u7801", ' ', /*#__PURE__*/_react["default"].createElement("a", {
      href: "javascript:void(0)",
      onClick: handleDownloadSources
    }, "\u5BFC\u51FA/\u4E0B\u8F7D zip \u5305")),
    key: "sources"
  }, gravityCode != null && /*#__PURE__*/_react["default"].createElement(_SourcesView.SourcesView, {
    height: sourcesViewHeight,
    code: gravityCode,
    onCodeChange: setGravityCode
  })), /*#__PURE__*/_react["default"].createElement(_next.Collapse.Panel, {
    title: /*#__PURE__*/_react["default"].createElement("span", null, "\u5728\u7EBF\u9884\u89C8", ' ', /*#__PURE__*/_react["default"].createElement("a", {
      href: "#refresh",
      onClick: function onClick(e) {
        e.preventDefault();
        e.stopPropagation();
        setRefresh(Date.now());
      }
    }, "\u5237\u65B0")),
    key: "preview"
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: "code-gen-result-gravity-demo",
    style: {
      height: gravityDemoHeight
    }
  }, /*#__PURE__*/_react["default"].createElement(_CodeGenPreview.CodeGenPreview, {
    code: gravityCode,
    height: gravityDemoHeight,
    refresh: refresh
  })))));
}

function convertCodeGenResult(result, schema) {
  var schemaFiles = {
    '/.project-schema.json': {
      fpath: '/.project-schema.json',
      code: JSON.stringify(schema, null, 2) + "\n"
    }
  };

  if (!result || !Array.isArray(result) || !result.length) {
    return {
      type: 'demo',
      modules: schemaFiles
    };
  }

  var code = {
    type: 'demo',
    modules: result.reduce(function (acc, file) {
      var _extends2;

      return (0, _extends3["default"])({}, acc, (_extends2 = {}, _extends2["/" + file.pathName] = {
        fpath: "/" + file.pathName,
        code: file.content,
        entry: undefined,
        packagejson: ['package.json'].includes(file.pathName) ? 1 : undefined
      }, _extends2));
    }, {})
  };
  var foundEntry = false; // 设置入口文件

  ['index.js', 'index.ts', 'index.tsx', 'app.js', 'app.ts', 'app.tsx'].forEach(function (fileName) {
    if (!foundEntry) {
      var filePath = "/src/" + fileName;

      if (code.modules[filePath]) {
        foundEntry = true;

        if (fileName === 'index.js') {
          code.modules[filePath].entry = 1;
        } else {
          code.modules['/src/index.js'] = {
            fpath: '/src/index.js',
            entry: 1,
            code: "import \"./" + fileName.replace(/\.\w+$/, '') + "\""
          };
        }
      }
    }
  });

  if (!foundEntry) {
    console.warn('Failed to find entry file for demo.');
  } // 补充 schema 文件


  Object.assign(code.modules, schemaFiles);
  return code;
}
