// @formily/next/esm/select-table/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import React, { useState, useMemo } from 'react';
import { observer, useFieldSchema, useField, RecursionField, } from '@formily/react';
import cls from 'classnames';
import { isArr, isBool, isFn } from '@formily/shared';
import { Search, Table } from '@alifd/next';
import { useFilterOptions } from './useFilterOptions';
import { useFlatOptions } from './useFlatOptions';
import { useSize } from './useSize';
import { useTitleAddon } from './useTitleAddon';
import { useCheckSlackly, getIndeterminate } from './useCheckSlackly';
import { getUISelected, getOutputData } from './utils';
import { usePrefixCls } from '../__builtins__';
var isColumnComponent = function (schema) {
    var _a;
    return ((_a = schema['x-component']) === null || _a === void 0 ? void 0 : _a.indexOf('Column')) > -1;
};
var useSources = function () {
    var arrayField = useField();
    var schema = useFieldSchema();
    var parseSources = function (schema) {
        var _a, _b, _c;
        if (isColumnComponent(schema)) {
            if (!((_a = schema['x-component-props']) === null || _a === void 0 ? void 0 : _a['dataIndex']) && !schema['name'])
                return [];
            var name_1 = ((_b = schema['x-component-props']) === null || _b === void 0 ? void 0 : _b['dataIndex']) || schema['name'];
            var field = arrayField.query(arrayField.address.concat(name_1)).take();
            var columnProps = ((_c = field === null || field === void 0 ? void 0 : field.component) === null || _c === void 0 ? void 0 : _c[1]) || schema['x-component-props'] || {};
            var display = (field === null || field === void 0 ? void 0 : field.display) || schema['x-display'];
            return [
                {
                    name: name_1,
                    display: display,
                    field: field,
                    schema: schema,
                    columnProps: __assign({ title: (field === null || field === void 0 ? void 0 : field.title) || columnProps.title }, columnProps),
                },
            ];
        }
        else if (schema.properties) {
            return schema.reduceProperties(function (buf, schema) {
                return buf.concat(parseSources(schema));
            }, []);
        }
    };
    var parseArrayItems = function (schema) {
        if (!schema)
            return [];
        var sources = [];
        var items = isArr(schema) ? schema : [schema];
        return items.reduce(function (columns, schema) {
            var item = parseSources(schema);
            if (item) {
                return columns.concat(item);
            }
            return columns;
        }, sources);
    };
    var validSchema = ((schema === null || schema === void 0 ? void 0 : schema.type) === 'array' && (schema === null || schema === void 0 ? void 0 : schema.items) ? schema.items : schema);
    return parseArrayItems(validSchema);
};
var useColumns = function (sources) {
    return sources.reduce(function (buf, _a, key) {
        var name = _a.name, columnProps = _a.columnProps, schema = _a.schema, display = _a.display;
        if (display !== 'visible')
            return buf;
        if (!isColumnComponent(schema))
            return buf;
        return buf.concat(__assign(__assign({}, columnProps), { key: key, dataIndex: name }));
    }, []);
};
var addPrimaryKey = function (dataSource, rowKey, primaryKey) {
    return dataSource.map(function (item) {
        var _a;
        var children = isArr(item.children)
            ? addPrimaryKey(item.children, rowKey, primaryKey)
            : {};
        return __assign(__assign(__assign({}, item), children), (_a = {}, _a[primaryKey] = rowKey(item), _a));
    });
};
export var SelectTable = observer(function (props) {
    var _a;
    var mode = props.mode, propsDataSource = props.dataSource, optionAsValue = props.optionAsValue, valueType = props.valueType, showSearch = props.showSearch, filterOption = props.filterOption, filterSort = props.filterSort, onSearch = props.onSearch, searchProps = props.searchProps, className = props.className, value = props.value, onChange = props.onChange, rowSelection = props.rowSelection, rowKey = props.primaryKey, otherTableProps = __rest(props, ["mode", "dataSource", "optionAsValue", "valueType", "showSearch", "filterOption", "filterSort", "onSearch", "searchProps", "className", "value", "onChange", "rowSelection", "primaryKey"]);
    var prefixCls = usePrefixCls('formily-select-table', props);
    var _b = __read(useState(), 2), searchValue = _b[0], setSearchValue = _b[1];
    var field = useField();
    var loading = isBool(props.loading) ? props.loading : field.loading;
    var disabled = field.disabled;
    var readOnly = field.readOnly;
    var readPretty = field.readPretty;
    var _c = useSize((_a = field.decoratorProps) === null || _a === void 0 ? void 0 : _a.size, searchProps === null || searchProps === void 0 ? void 0 : searchProps.size, props === null || props === void 0 ? void 0 : props.size), searchSize = _c.searchSize, tableSize = _c.tableSize;
    var primaryKey = isFn(rowKey) ? '__formily_key__' : rowKey;
    var sources = useSources();
    var columns = useColumns(sources);
    // dataSource
    var dataSource = isArr(propsDataSource) ? propsDataSource : field.dataSource;
    dataSource = isFn(rowKey)
        ? addPrimaryKey(dataSource, rowKey, primaryKey)
        : dataSource;
    // Filter dataSource By Search
    var filteredDataSource = useFilterOptions(dataSource, searchValue, filterOption, rowSelection === null || rowSelection === void 0 ? void 0 : rowSelection.checkStrictly);
    // Order dataSource By filterSort
    var orderedFilteredDataSource = useMemo(function () {
        if (!filterSort) {
            return filteredDataSource;
        }
        return __spreadArray([], __read(filteredDataSource), false).sort(function (a, b) { return filterSort(a, b); });
    }, [filteredDataSource, filterSort]);
    var flatDataSource = useFlatOptions(dataSource);
    var flatFilteredDataSource = useFlatOptions(filteredDataSource);
    // 分页或异步查询时，dataSource会丢失已选数据，配置optionAsValue则无法获取已选数据，需要进行合并
    var getWholeDataSource = function () {
        if (optionAsValue && mode === 'multiple' && (value === null || value === void 0 ? void 0 : value.length)) {
            var map_1 = new Map();
            var arr = __spreadArray(__spreadArray([], __read(flatDataSource), false), __read(value), false);
            arr.forEach(function (item) {
                if (!map_1.has(item[primaryKey])) {
                    map_1.set(item[primaryKey], item);
                }
            });
            return __spreadArray([], __read(map_1.values()), false);
        }
        return flatDataSource;
    };
    // selected keys for Table UI
    var selected = getUISelected(value, flatDataSource, primaryKey, valueType, optionAsValue, mode, rowSelection === null || rowSelection === void 0 ? void 0 : rowSelection.checkStrictly, rowKey);
    // readPretty Value
    var readPrettyDataSource = useFilterOptions(orderedFilteredDataSource, selected, function (value, item) { return value.includes(item[primaryKey]); });
    var onInnerSearch = function (searchText) {
        var formatted = (searchText || '').trim();
        setSearchValue(searchText);
        onSearch === null || onSearch === void 0 ? void 0 : onSearch(formatted);
    };
    var onInnerChange = function (selectedRowKeys) {
        if (readOnly) {
            return;
        }
        // 筛选后onChange默认的records数据不完整，此处需使用完整数据过滤
        var wholeRecords = getWholeDataSource().filter(function (item) {
            return selectedRowKeys.includes(item === null || item === void 0 ? void 0 : item[primaryKey]);
        });
        var _a = getOutputData(selectedRowKeys, wholeRecords, dataSource, primaryKey, valueType, optionAsValue, mode, rowSelection === null || rowSelection === void 0 ? void 0 : rowSelection.checkStrictly), outputValue = _a.outputValue, outputOptions = _a.outputOptions;
        onChange === null || onChange === void 0 ? void 0 : onChange(outputValue, outputOptions);
    };
    var onRowClick = function (record) {
        if (readPretty || disabled || readOnly || (record === null || record === void 0 ? void 0 : record.disabled)) {
            return;
        }
        var selectedRowKey = record === null || record === void 0 ? void 0 : record[primaryKey];
        var isSelected = selected === null || selected === void 0 ? void 0 : selected.includes(selectedRowKey);
        var selectedRowKeys = [];
        if (mode === 'single') {
            selectedRowKeys = [selectedRowKey];
        }
        else {
            if (isSelected) {
                selectedRowKeys = selected.filter(function (item) { return item !== selectedRowKey; });
            }
            else {
                selectedRowKeys = __spreadArray(__spreadArray([], __read(selected), false), [selectedRowKey], false);
            }
        }
        if ((rowSelection === null || rowSelection === void 0 ? void 0 : rowSelection.checkStrictly) !== false) {
            onInnerChange(selectedRowKeys);
        }
        else {
            onSlacklyChange(selectedRowKeys);
        }
    };
    // TreeData SlacklyChange
    var onSlacklyChange = function (currentSelected) {
        var selectedRowKeys = useCheckSlackly(currentSelected, selected, flatDataSource, flatFilteredDataSource, primaryKey, rowSelection === null || rowSelection === void 0 ? void 0 : rowSelection.checkStrictly).selectedRowKeys;
        onInnerChange(selectedRowKeys);
    };
    // Table All Checkbox
    var titleAddon = useTitleAddon(selected, flatDataSource, flatFilteredDataSource, primaryKey, mode, disabled, readOnly, rowSelection === null || rowSelection === void 0 ? void 0 : rowSelection.checkStrictly, onInnerChange);
    return (React.createElement("div", { className: prefixCls },
        showSearch ? (React.createElement(Search, __assign({}, searchProps, { className: cls("".concat(prefixCls, "-search"), searchProps === null || searchProps === void 0 ? void 0 : searchProps.className), style: __assign({ width: '100%' }, searchProps === null || searchProps === void 0 ? void 0 : searchProps.style), onSearch: onInnerSearch, onChange: onInnerSearch, disabled: disabled, readOnly: readOnly, size: searchSize, buttonProps: __assign(__assign({}, searchProps === null || searchProps === void 0 ? void 0 : searchProps.buttonProps), { loading: loading }) }))) : null,
        React.createElement(Table, __assign({}, otherTableProps, { className: cls("".concat(prefixCls, "-table"), className), dataSource: readPretty ? readPrettyDataSource : orderedFilteredDataSource, rowSelection: readPretty
                ? undefined
                : __assign(__assign(__assign({}, rowSelection), titleAddon), { getProps: function (record, index) {
                        var _a;
                        return (__assign(__assign(__assign({}, (_a = rowSelection === null || rowSelection === void 0 ? void 0 : rowSelection.getProps) === null || _a === void 0 ? void 0 : _a.call(rowSelection, record, index)), ((rowSelection === null || rowSelection === void 0 ? void 0 : rowSelection.checkStrictly) !== false
                            ? {}
                            : {
                                indeterminate: getIndeterminate(record, flatDataSource, selected, primaryKey),
                            })), { disabled: disabled || (record === null || record === void 0 ? void 0 : record.disabled) }));
                    }, selectedRowKeys: selected, onChange: (rowSelection === null || rowSelection === void 0 ? void 0 : rowSelection.checkStrictly) !== false
                        ? onInnerChange
                        : onSlacklyChange, mode: mode }), columns: props.columns || columns, primaryKey: primaryKey, loading: loading, size: tableSize, onRowClick: function (record, index, e) {
                var _a;
                // fusion
                onRowClick(record);
                (_a = props.onRowClick) === null || _a === void 0 ? void 0 : _a.call(props, record, index, e);
            } }), ''),
        sources.map(function (column, key) {
            //专门用来承接对Column的状态管理
            if (!isColumnComponent(column.schema))
                return;
            return React.createElement(RecursionField, {
                name: column.name,
                schema: column.schema,
                onlyRenderSelf: true,
                key: key,
            });
        })));
});
var TableColumn = function () { return React.createElement(React.Fragment, null); };
SelectTable.Column = TableColumn;
SelectTable.defaultProps = {
    showSearch: false,
    valueType: 'all',
    primaryKey: 'key',
    mode: 'multiple',
};
export default SelectTable;
