// @alilc/lowcode-plugin-code-generator/lib/components/CodeSandboxPreview/CodeSandboxPreview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.CodeSandboxPreview = CodeSandboxPreview;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _next = require("@alifd/next");

var _fixPreviewCode = require("../CodeGenPreview/fixPreviewCode");

require("./CodeSandboxPreview.scss");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// 使用 CodeSandbox 来进行预览
// @see https://codesandbox.io/docs/api#get-request
function CodeSandboxPreview(_ref) {
  var code = _ref.code,
      height = _ref.height;
  var parameters = (0, _react.useMemo)(function () {
    var files = {};

    if (code && code.modules) {
      var fixedCode = (0, _fixPreviewCode.fixPreviewCode)(code);
      Object.values(fixedCode.modules).forEach(function (file) {
        var _file$fpath;

        files[file === null || file === void 0 ? void 0 : (_file$fpath = file.fpath) === null || _file$fpath === void 0 ? void 0 : _file$fpath.slice(1)] = {
          isBinary: false,
          content: file.code
        };
      }); // 入口文件需要顺便引入下样式

      files['src/index.js'] = {
        isBinary: false,
        content: "\n// \u76EE\u524D\u9700\u8981\u5355\u72EC\u5F15\u5165\u4E0B\u6837\u5F0F\u6587\u4EF6\nimport \"@alifd/next/dist/next.css\";\nimport \"@alifd/pro-layout/dist/AlifdProLayout.css\";\n\n// \u5F15\u5165\u5165\u53E3\u6587\u4EF6\nimport './app';\n"
      };
    }

    return {
      files: files,
      template: 'create-react-app'
    };
  }, [code]);

  var _useState = (0, _react.useState)({
    parameters: parameters,
    sandboxId: '',
    isCreating: false,
    hasError: false,
    error: null
  }),
      state = _useState[0],
      setState = _useState[1];

  (0, _react.useEffect)(function () {
    if (state.parameters !== parameters || state.sandboxId === 'retry' || !state.sandboxId) {
      var hasCanceled = false;
      setState(function (prev) {
        return (0, _extends2["default"])({}, prev, {
          hasError: false,
          code: code,
          parameters: parameters,
          isCreating: true
        });
      });
      (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var sandboxId;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.prev = 0;
                _context.next = 3;
                return createCodeSandbox(parameters);

              case 3:
                sandboxId = _context.sent;

                if (!hasCanceled) {
                  setState(function (prev) {
                    return (0, _extends2["default"])({}, prev, {
                      hasError: false,
                      isCreating: false,
                      sandboxId: sandboxId
                    });
                  });
                }

                _context.next = 10;
                break;

              case 7:
                _context.prev = 7;
                _context.t0 = _context["catch"](0);

                if (!hasCanceled) {
                  setState(function (prev) {
                    return (0, _extends2["default"])({}, prev, {
                      hasError: true,
                      error: _context.t0,
                      isCreating: false
                    });
                  });
                }

              case 10:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, null, [[0, 7]]);
      }))();
      return function () {
        hasCanceled = true;
      };
    }

    return function () {};
  }, [parameters, state.sandboxId, state.parameters]);

  var handleRetry = function handleRetry() {
    setState(function (prev) {
      return (0, _extends2["default"])({}, prev, {
        hasError: false,
        sandboxId: 'retry'
      });
    });
  };

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: "code-gen-plugin-code-sandbox-preview",
    style: {
      height: height
    },
    "data-code-sandbox-id": state.sandboxId
  }, function () {
    if (state.hasError) {
      return /*#__PURE__*/_react["default"].createElement(_next.Message, {
        title: "\u751F\u6210 CodeSandbox \u9884\u89C8\u5E94\u7528\u5931\u8D25"
      }, /*#__PURE__*/_react["default"].createElement("p", null, "\u8BE6\u7EC6\u9519\u8BEF\uFF1A", "" + (state.error || '网络开小差了')), /*#__PURE__*/_react["default"].createElement("p", null, /*#__PURE__*/_react["default"].createElement(_next.Button, {
        onClick: handleRetry
      }, "\u91CD\u65B0\u5C1D\u8BD5\u4E0B")));
    }

    return state.sandboxId ? /*#__PURE__*/_react["default"].createElement("iframe", {
      src: "https://codesandbox.io/embed/" + state.sandboxId + "?autoresize=1&fontsize=14&hidenavigation=1&theme=dark&view=preview",
      title: "CodeSandbox Preview",
      allow: "accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking",
      sandbox: "allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
    }) : /*#__PURE__*/_react["default"].createElement(_next.Loading, {
      tip: "\u6B63\u5728\u751F\u6210 CodeSandbox \u9884\u89C8\u5E94\u7528..."
    });
  }());
}

function createCodeSandbox(_x) {
  return _createCodeSandbox.apply(this, arguments);
}

function _createCodeSandbox() {
  _createCodeSandbox = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(parameters) {
    var res, json, _ref3, sandbox_id;

    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            if (Object.entries((parameters === null || parameters === void 0 ? void 0 : parameters.files) || {}).length) {
              _context2.next = 2;
              break;
            }

            return _context2.abrupt("return", '');

          case 2:
            _context2.next = 4;
            return fetch('https://codesandbox.io/api/v1/sandboxes/define?json=1', {
              method: 'POST',
              headers: {
                'Content-Type': 'application/json',
                Accept: 'application/json'
              },
              body: JSON.stringify(parameters)
            });

          case 4:
            res = _context2.sent;

            if (res.ok) {
              _context2.next = 7;
              break;
            }

            throw new Error("\u521B\u5EFA CodeSandbox \u5931\u8D25\uFF0C\u9519\u8BEF\u7801\uFF1A" + res.status + " " + res.statusText);

          case 7:
            _context2.next = 9;
            return res.json()["catch"](function (err) {
              throw new Error("\u521B\u5EFA CodeSandbox \u5931\u8D25\uFF0C\u670D\u52A1\u5F02\u5E38(" + ((err === null || err === void 0 ? void 0 : err.message) || err || '未知异常') + ")");
            });

          case 9:
            json = _context2.sent;
            _ref3 = json || {}, sandbox_id = _ref3.sandbox_id;

            if (!(!sandbox_id || typeof sandbox_id !== 'string')) {
              _context2.next = 13;
              break;
            }

            throw new Error("\u521B\u5EFA CodeSandbox \u5931\u8D25\uFF0C\u670D\u52A1\u54CD\u5E94\u5F02\u5E38");

          case 13:
            return _context2.abrupt("return", sandbox_id);

          case 14:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _createCodeSandbox.apply(this, arguments);
}
