// @alilc/lowcode-designer/lib/designer/setting/setting-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.SettingField = void 0;
exports.isSettingField = isSettingField;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));
var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));
var _utils = require("./utils");
var _settingPropEntry = require("./setting-prop-entry");
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _lowcodeUtils = require("@alilc/lowcode-utils");
var _excluded = ["title", "items", "setter", "extraProps"];
var _dec, _class, _descriptor;
function getSettingFieldCollectorKey(parent, config) {
  var cur = parent;
  var path = [config.name];
  while (cur !== parent.top) {
    if (cur instanceof SettingField && cur.type !== 'group') {
      path.unshift(cur.name);
    }
    cur = cur.parent;
  }
  return path.join('.');
}
var SettingField = exports.SettingField = (_dec = _lowcodeEditorCore.obx.ref, (_class = /*#__PURE__*/function (_SettingPropEntry) {
  (0, _inheritsLoose2["default"])(SettingField, _SettingPropEntry);
  function SettingField(parent, config, settingFieldCollector) {
    var _this;
    _this = _SettingPropEntry.call(this, parent, config.name, config.type) || this;
    _this.settingFieldCollector = settingFieldCollector;
    _this.isSettingField = true;
    _this.isRequired = void 0;
    _this.transducer = void 0;
    _this._config = void 0;
    _this.hotValue = void 0;
    _this.parent = void 0;
    _this.extraProps = void 0;
    // ==== dynamic properties ====
    _this._title = void 0;
    _this._setter = void 0;
    (0, _initializerDefineProperty2["default"])(_this, "_expanded", _descriptor, (0, _assertThisInitialized2["default"])(_this));
    _this._items = [];
    (0, _lowcodeEditorCore.makeObservable)((0, _assertThisInitialized2["default"])(_this));
    var title = config.title,
      items = config.items,
      setter = config.setter,
      extraProps = config.extraProps,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(config, _excluded);
    _this.parent = parent;
    _this._config = config;
    _this._title = title;
    _this._setter = setter;
    _this.extraProps = (0, _extends2["default"])({}, rest, extraProps);
    _this.isRequired = config.isRequired || (setter === null || setter === void 0 ? void 0 : setter.isRequired);
    _this._expanded = !(extraProps !== null && extraProps !== void 0 && extraProps.defaultCollapsed);

    // initial items
    if (items && items.length > 0) {
      _this.initItems(items, settingFieldCollector);
    }
    if (_this.type !== 'group' && settingFieldCollector && config.name) {
      settingFieldCollector(getSettingFieldCollectorKey(parent, config), (0, _assertThisInitialized2["default"])(_this));
    }

    // compatiable old config
    _this.transducer = new _utils.Transducer((0, _assertThisInitialized2["default"])(_this), {
      setter: setter
    });
    return _this;
  }
  var _proto = SettingField.prototype;
  _proto.setExpanded = function setExpanded(value) {
    this._expanded = value;
  };
  _proto.initItems = function initItems(items, settingFieldCollector) {
    var _this2 = this;
    this._items = items.map(function (item) {
      if ((0, _lowcodeUtils.isCustomView)(item)) {
        return item;
      }
      return new SettingField(_this2, item, settingFieldCollector);
    });
  };
  _proto.disposeItems = function disposeItems() {
    this._items.forEach(function (item) {
      return isSettingField(item) && item.purge();
    });
    this._items = [];
  }

  // 创建子配置项，通常用于 object/array 类型数据
  ;
  _proto.createField = function createField(config) {
    var _this$settingFieldCol;
    (_this$settingFieldCol = this.settingFieldCollector) === null || _this$settingFieldCol === void 0 ? void 0 : _this$settingFieldCol.call(this, getSettingFieldCollectorKey(this.parent, config), this);
    return new SettingField(this, config, this.settingFieldCollector);
  };
  _proto.purge = function purge() {
    this.disposeItems();
  }

  // ======= compatibles for vision ======
  ;
  _proto.getConfig = function getConfig(configName) {
    if (configName) {
      return this.config[configName];
    }
    return this._config;
  };
  _proto.getItems = function getItems(filter) {
    return this._items.filter(function (item) {
      if (filter) {
        return filter(item);
      }
      return true;
    });
  };
  _proto.setValue = function setValue(val, isHotValue, force, extraOptions) {
    if (isHotValue) {
      this.setHotValue(val, extraOptions);
      return;
    }
    _SettingPropEntry.prototype.setValue.call(this, val, false, false, extraOptions);
  };
  _proto.getHotValue = function getHotValue() {
    if (this.hotValue) {
      return this.hotValue;
    }
    // avoid View modify
    var v = (0, _lowcodeUtils.cloneDeep)(this.getMockOrValue());
    if (v == null) {
      v = this.extraProps.defaultValue;
    }
    return this.transducer.toHot(v);
  }

  /* istanbul ignore next */;
  _proto.setMiniAppDataSourceValue = function setMiniAppDataSourceValue(data, options) {
    this.hotValue = data;
    var v = this.transducer.toNative(data);
    this.setValue(v, false, false, options);
    // dirty fix list setter
    if (Array.isArray(data) && data[0] && data[0].__sid__) {
      return;
    }
    this.valueChange();
  };
  _proto.setHotValue = function setHotValue(data, options) {
    this.hotValue = data;
    var value = this.transducer.toNative(data);
    if (options) {
      options.fromSetHotValue = true;
    } else {
      options = {
        fromSetHotValue: true
      };
    }
    if (this.isUseVariable()) {
      var oldValue = this.getValue();
      if ((0, _lowcodeUtils.isJSExpression)(value)) {
        this.setValue({
          type: 'JSExpression',
          value: value.value,
          mock: oldValue.mock
        }, false, false, options);
      } else {
        this.setValue({
          type: 'JSExpression',
          value: oldValue.value,
          mock: value
        }, false, false, options);
      }
    } else {
      this.setValue(value, false, false, options);
    }

    // dirty fix list setter
    if (Array.isArray(data) && data[0] && data[0].__sid__) {
      return;
    }
    this.valueChange(options);
  };
  _proto.onEffect = function onEffect(action) {
    return this.designer.autorun(action, true);
  };
  _proto.internalToShellField = function internalToShellField() {
    return this.designer.shellModelFactory.createSettingField(this);
  };
  (0, _createClass2["default"])(SettingField, [{
    key: "title",
    get: function get() {
      return this._title || (typeof this.name === 'number' ? (0, _lowcodeEditorCore.intl)('Item') + " " + this.name : this.name);
    }
  }, {
    key: "setter",
    get: function get() {
      var _this3 = this;
      if (!this._setter) {
        return null;
      }
      if ((0, _lowcodeUtils.isDynamicSetter)(this._setter)) {
        return (0, _lowcodeEditorCore.untracked)(function () {
          var _this3$_setter;
          var shellThis = _this3.internalToShellField();
          return (_this3$_setter = _this3._setter) === null || _this3$_setter === void 0 ? void 0 : _this3$_setter.call(shellThis, shellThis);
        });
      }
      return this._setter;
    }
  }, {
    key: "expanded",
    get: function get() {
      return this._expanded;
    }
  }, {
    key: "items",
    get: function get() {
      return this._items;
    }
  }, {
    key: "config",
    get: function get() {
      return this._config;
    }
  }]);
  return SettingField;
}(_settingPropEntry.SettingPropEntry), (_descriptor = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "_expanded", [_dec], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return true;
  }
}), (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "setter", [_lowcodeEditorCore.computed], Object.getOwnPropertyDescriptor(_class.prototype, "setter"), _class.prototype), (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "setValue", [_lowcodeEditorCore.action], Object.getOwnPropertyDescriptor(_class.prototype, "setValue"), _class.prototype), (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "setMiniAppDataSourceValue", [_lowcodeEditorCore.action], Object.getOwnPropertyDescriptor(_class.prototype, "setMiniAppDataSourceValue"), _class.prototype), (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "setHotValue", [_lowcodeEditorCore.action], Object.getOwnPropertyDescriptor(_class.prototype, "setHotValue"), _class.prototype)), _class));
/**
 * @deprecated use same function from '@alilc/lowcode-utils' instead
 */
function isSettingField(obj) {
  return obj && obj.isSettingField;
}
