// @alifd/meet-react/es/select-drawer/base.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
import { __rest } from "tslib";
import classNames from 'classnames';
import React, { createElement, useEffect, useState } from "react";
import Button from '../button';
import Drawer from '../drawer';
import View from '../view';
import { useLocale } from '../locale';
import defaultLang from '../locale/lang/zh-cn';
export default (function (props) {
  var _props$prefix = props.prefix,
    prefix = _props$prefix === void 0 ? 'mt-' : _props$prefix,
    _props$visible = props.visible,
    visible = _props$visible === void 0 ? false : _props$visible,
    drawerClassName = props.drawerClassName,
    drawerTitle = props.drawerTitle,
    onCancel = props.onCancel,
    onClose = props.onClose,
    onOk = props.onOk,
    children = props.children,
    others = __rest(props, ["prefix", "visible", "drawerClassName", "drawerTitle", "onCancel", "onClose", "onOk", "children"]);
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    drawerVisible = _useState2[0],
    setDrawerVisible = _useState2[1];
  var clsPrefix = "".concat(prefix, "selectdrawer");
  useEffect(function () {
    if (drawerVisible !== visible) {
      setDrawerVisible(visible);
    }
  }, [visible]);
  var locale = useLocale('SelectDrawer', others, defaultLang.SelectDrawer);
  return /*#__PURE__*/React.createElement(Drawer, _extends({}, others, {
    visible: drawerVisible,
    placement: "bottom",
    closeMode: ['mask'],
    onClose: onClose,
    className: classNames("".concat(clsPrefix, "-drawer"), drawerClassName)
  }), /*#__PURE__*/React.createElement(View, {
    className: "".concat(clsPrefix, "-toolbar")
  }, /*#__PURE__*/React.createElement(Button, {
    model: "text",
    onClick: onCancel
  }, locale.cancel), /*#__PURE__*/React.createElement(View, {
    className: "".concat(clsPrefix, "-toolbar-title")
  }, drawerTitle), /*#__PURE__*/React.createElement(Button, {
    model: "text",
    type: "primary",
    onClick: onOk
  }, locale.ok)), /*#__PURE__*/React.createElement(View, {
    className: "".concat(clsPrefix, "-drawer-content")
  }, children));
});
