// @alilc/lowcode-react-renderer/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports["default"] = void 0;
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = _interopRequireWildcard(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _lowcodeRendererCore = require("@alilc/lowcode-renderer-core");
var _configProvider = _interopRequireDefault(require("@alifd/next/lib/config-provider"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
window.React = _react["default"];
window.ReactDom = _reactDom["default"];
_lowcodeRendererCore.adapter.setRuntime({
  Component: _react.Component,
  PureComponent: _react.PureComponent,
  createContext: _react.createContext,
  createElement: _react.createElement,
  forwardRef: _react.forwardRef,
  findDOMNode: _reactDom["default"].findDOMNode
});
_lowcodeRendererCore.adapter.setRenderers({
  PageRenderer: (0, _lowcodeRendererCore.pageRendererFactory)(),
  ComponentRenderer: (0, _lowcodeRendererCore.componentRendererFactory)(),
  BlockRenderer: (0, _lowcodeRendererCore.blockRendererFactory)(),
  AddonRenderer: (0, _lowcodeRendererCore.addonRendererFactory)(),
  TempRenderer: (0, _lowcodeRendererCore.tempRendererFactory)(),
  DivRenderer: (0, _lowcodeRendererCore.blockRendererFactory)()
});
_lowcodeRendererCore.adapter.setConfigProvider(_configProvider["default"]);
function factory() {
  var Renderer = (0, _lowcodeRendererCore.rendererFactory)();
  return /*#__PURE__*/function (_Renderer) {
    (0, _inheritsLoose2["default"])(ReactRenderer, _Renderer);
    function ReactRenderer(props, context) {
      var _this;
      _this = _Renderer.call(this, props, context) || this;
      _this.props = void 0;
      _this.context = void 0;
      _this.setState = void 0;
      _this.forceUpdate = void 0;
      _this.refs = void 0;
      return _this;
    }
    var _proto = ReactRenderer.prototype;
    _proto.isValidComponent = function isValidComponent(obj) {
      var _obj$prototype;
      return (obj === null || obj === void 0 ? void 0 : (_obj$prototype = obj.prototype) === null || _obj$prototype === void 0 ? void 0 : _obj$prototype.isReactComponent) || (obj === null || obj === void 0 ? void 0 : obj.prototype) instanceof _react.Component;
    };
    return ReactRenderer;
  }(Renderer);
}
var _default = exports["default"] = factory();
