// @alilc/lowcode-designer/lib/project/project-view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.ProjectView = exports.BuiltinLoading = void 0;
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = require("react");
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _builtinSimulator = require("../builtin-simulator");
require("./project.less");
var _class;
var BuiltinLoading = exports.BuiltinLoading = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(BuiltinLoading, _Component);
  function BuiltinLoading() {
    return _Component.apply(this, arguments) || this;
  }
  var _proto = BuiltinLoading.prototype;
  _proto.render = function render() {
    return /*#__PURE__*/React.createElement("div", {
      id: "engine-loading-wrapper"
    }, /*#__PURE__*/React.createElement("img", {
      width: "154",
      height: "100",
      src: "https://img.alicdn.com/tfs/TB1CmVgayERMeJjy0FcXXc7opXa-308-200.gif"
    }));
  };
  return BuiltinLoading;
}(_react.Component);
var ProjectView = exports.ProjectView = (0, _lowcodeEditorCore.observer)(_class = /*#__PURE__*/function (_Component2) {
  (0, _inheritsLoose2["default"])(ProjectView, _Component2);
  function ProjectView() {
    return _Component2.apply(this, arguments) || this;
  }
  var _proto2 = ProjectView.prototype;
  _proto2.componentDidMount = function componentDidMount() {
    var _this = this;
    var designer = this.props.designer;
    var project = designer.project;
    project.onRendererReady(function () {
      _this.forceUpdate();
    });
  };
  _proto2.render = function render() {
    var _project$simulator;
    var designer = this.props.designer;
    var project = designer.project,
      simulatorProps = designer.projectSimulatorProps;
    var Simulator = designer.simulatorComponent || _builtinSimulator.BuiltinSimulatorHostView;
    var Loading = _lowcodeEditorCore.engineConfig.get('loadingComponent', BuiltinLoading);
    return /*#__PURE__*/React.createElement("div", {
      className: "lc-project"
    }, /*#__PURE__*/React.createElement("div", {
      className: "lc-simulator-shell"
    }, !(project !== null && project !== void 0 && (_project$simulator = project.simulator) !== null && _project$simulator !== void 0 && _project$simulator.renderer) && /*#__PURE__*/React.createElement(Loading, null), /*#__PURE__*/React.createElement(Simulator, simulatorProps)));
  };
  return ProjectView;
}(_react.Component)) || _class;
