// @alilc/lowcode-engine/lib/modules/live-editing.js
"use strict";

exports.__esModule = true;
exports.liveEditingRule = liveEditingRule;
exports.liveEditingSaveHander = void 0;
var _lowcodeDesigner = require("@alilc/lowcode-designer");
var _lowcodeUtils = require("@alilc/lowcode-utils");
function getText(node, prop) {
  var p = node.getProp(prop, false);
  if (!p || p.isUnset()) {
    return null;
  }
  var v = p.getValue();
  if ((0, _lowcodeUtils.isJSExpression)(v)) {
    v = v.mock;
  }
  if (v == null) {
    return null;
  }
  if (p.type === 'literal') {
    return v;
  }
  return Symbol["for"]('not-literal');
}
function liveEditingRule(target) {
  // for vision components specific
  var node = target.node,
    event = target.event;
  var targetElement = event.target;
  if (!Array.from(targetElement.childNodes).every(function (item) {
    return item.nodeType === Node.TEXT_NODE;
  })) {
    return null;
  }
  var innerText = targetElement.innerText;
  var propTarget = ['title', 'label', 'text', 'content', 'children'].find(function (prop) {
    return equalText(getText(node, prop), innerText);
  });
  if (propTarget) {
    return {
      propElement: targetElement,
      propTarget: propTarget
    };
  }
  return null;
}
function equalText(v, innerText) {
  // TODO: enhance compare text logic
  if (typeof v !== 'string') {
    return false;
  }
  return v.trim() === innerText;
}
var liveEditingSaveHander = exports.liveEditingSaveHander = {
  condition: function condition(prop) {
    return prop.type === 'expression';
  },
  onSaveContent: function onSaveContent(content, prop) {
    var v = prop.getValue();
    var data = v;
    if ((0, _lowcodeUtils.isJSExpression)(v)) {
      data = v.mock;
    }
    data = content;
    if ((0, _lowcodeUtils.isJSExpression)(v)) {
      prop.setValue({
        type: 'JSExpression',
        value: v.value,
        mock: data
      });
    } else {
      prop.setValue(data);
    }
  }
};
// TODO:
// 非文本编辑
//  国际化数据，改变当前
//  JSExpression, 改变 mock 或 弹出绑定变量

_lowcodeDesigner.LiveEditing.addLiveEditingSpecificRule(liveEditingRule);
_lowcodeDesigner.LiveEditing.addLiveEditingSaveHandler(liveEditingSaveHander);
