// @alilc/lowcode-engine/lib/modules/symbols.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;
var _lowcodeShell = require("@alilc/lowcode-shell");
var _default = exports["default"] = {
  projectSymbol: _lowcodeShell.projectSymbol,
  documentSymbol: _lowcodeShell.documentSymbol,
  nodeSymbol: _lowcodeShell.nodeSymbol,
  nodeChildrenSymbol: _lowcodeShell.nodeChildrenSymbol,
  skeletonSymbol: _lowcodeShell.skeletonSymbol,
  editorSymbol: _lowcodeShell.editorSymbol,
  designerSymbol: _lowcodeShell.designerSymbol,
  settingPropEntrySymbol: _lowcodeShell.settingFieldSymbol,
  settingTopEntrySymbol: _lowcodeShell.settingTopEntrySymbol,
  designerCabinSymbol: _lowcodeShell.designerCabinSymbol,
  editorCabinSymbol: _lowcodeShell.editorCabinSymbol,
  skeletonCabinSymbol: _lowcodeShell.skeletonCabinSymbol,
  propSymbol: _lowcodeShell.propSymbol,
  simulatorHostSymbol: _lowcodeShell.simulatorHostSymbol,
  skeletonItemSymbol: _lowcodeShell.skeletonItemSymbol,
  simulatorRenderSymbol: _lowcodeShell.simulatorRenderSymbol
};
