// universal-request/es/quickapp/index.js
import { isQuickApp } from 'universal-env';
import otherModule from '../index';
var exportModule;

function json2string(obj) {
  try {
    return JSON.stringify(obj);
  } catch (e) {
    return obj || '';
  }
}

function request(params) {
  if (!params.url || params.url === '') {
    throw 'lack of url';
  }

  var url = params.url,
      method = params.method,
      _params$data = params.data,
      data = _params$data === void 0 ? {} : _params$data,
      _params$dataType = params.dataType,
      dataType = _params$dataType === void 0 ? 'json' : _params$dataType,
      headers = params.headers;
  return new Promise(function (resolve, reject) {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    var fetch = require('@system.fetch');

    fetch.fetch({
      url: url,
      responseType: dataType,
      method: method,
      data: data,
      header: headers,
      success: function success(response) {
        resolve(response);
      },
      fail: function fail(data, code) {
        reject({
          code: code,
          message: json2string(data)
        });
      }
    });
  });
}

if (isQuickApp) {
  exportModule = request;
} else {
  exportModule = otherModule;
}

export default exportModule;
