// @alilc/lowcode-shell/lib/api/canvas.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.Canvas = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _lowcodeDesigner = require("@alilc/lowcode-designer");
var _symbols = require("../symbols");
var _model = require("../model");
var clipboardInstanceSymbol = Symbol('clipboardInstace');
var Canvas = exports.Canvas = /*#__PURE__*/function () {
  function Canvas(editor, workspaceMode) {
    if (workspaceMode === void 0) {
      workspaceMode = false;
    }
    this.workspaceMode = workspaceMode;
    this[_symbols.editorSymbol] = void 0;
    this[clipboardInstanceSymbol] = void 0;
    this[_symbols.editorSymbol] = editor;
    this[clipboardInstanceSymbol] = new _model.Clipboard();
  }
  var _proto = Canvas.prototype;
  _proto.createScrollTarget = function createScrollTarget(shell) {
    return new _lowcodeDesigner.ScrollTarget(shell);
  };
  _proto.createScroller = function createScroller(scrollable) {
    return this[_symbols.designerSymbol].createScroller(scrollable);
  }

  /**
   * 创建插入位置，考虑放到 dragon 中
   */;
  _proto.createLocation = function createLocation(locationData) {
    return new _model.DropLocation(this[_symbols.designerSymbol].createLocation((0, _extends2["default"])({}, locationData, {
      target: locationData.target[_symbols.nodeSymbol]
    })));
  }

  /**
   * @deprecated
   */;
  (0, _createClass2["default"])(Canvas, [{
    key: _symbols.designerSymbol,
    get: function get() {
      return this[_symbols.editorSymbol].get('designer');
    }
  }, {
    key: "dragon",
    get: function get() {
      return _model.Dragon.create(this[_symbols.designerSymbol].dragon, this.workspaceMode);
    }
  }, {
    key: "activeTracker",
    get: function get() {
      var activeTracker = new _model.ActiveTracker(this[_symbols.designerSymbol].activeTracker);
      return activeTracker;
    }
  }, {
    key: "isInLiveEditing",
    get: function get() {
      var _this$editorSymbol$ge, _this$editorSymbol$ge2, _this$editorSymbol$ge3, _this$editorSymbol$ge4;
      return Boolean((_this$editorSymbol$ge = this[_symbols.editorSymbol].get('designer')) === null || _this$editorSymbol$ge === void 0 ? void 0 : (_this$editorSymbol$ge2 = _this$editorSymbol$ge.project) === null || _this$editorSymbol$ge2 === void 0 ? void 0 : (_this$editorSymbol$ge3 = _this$editorSymbol$ge2.simulator) === null || _this$editorSymbol$ge3 === void 0 ? void 0 : (_this$editorSymbol$ge4 = _this$editorSymbol$ge3.liveEditing) === null || _this$editorSymbol$ge4 === void 0 ? void 0 : _this$editorSymbol$ge4.editing);
    }
  }, {
    key: "clipboard",
    get: function get() {
      return this[clipboardInstanceSymbol];
    }
  }, {
    key: "dropLocation",
    get: function get() {
      return _model.DropLocation.create(this[_symbols.designerSymbol].dropLocation || null);
    }
  }]);
  return Canvas;
}();
