// @alilc/lowcode-shell/lib/api/config.js
"use strict";

exports.__esModule = true;
exports.Config = void 0;
var _symbols = require("../symbols");
var Config = exports.Config = /*#__PURE__*/function () {
  function Config(innerEngineConfig) {
    this[_symbols.configSymbol] = void 0;
    this[_symbols.configSymbol] = innerEngineConfig;
  }
  var _proto = Config.prototype;
  _proto.has = function has(key) {
    return this[_symbols.configSymbol].has(key);
  };
  _proto.get = function get(key, defaultValue) {
    return this[_symbols.configSymbol].get(key, defaultValue);
  };
  _proto.set = function set(key, value) {
    this[_symbols.configSymbol].set(key, value);
  };
  _proto.setConfig = function setConfig(config) {
    this[_symbols.configSymbol].setConfig(config);
  };
  _proto.onceGot = function onceGot(key) {
    return this[_symbols.configSymbol].onceGot(key);
  };
  _proto.onGot = function onGot(key, fn) {
    return this[_symbols.configSymbol].onGot(key, fn);
  };
  _proto.getPreference = function getPreference() {
    return this[_symbols.configSymbol].getPreference();
  };
  return Config;
}();
