// @alilc/lowcode-shell/lib/api/material.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.Material = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _lowcodeDesigner = require("@alilc/lowcode-designer");
var _lowcodeUtils = require("@alilc/lowcode-utils");
var _symbols = require("../symbols");
var _model = require("../model");
function _createForOfIteratorHelperLoose(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (it) return (it = it.call(o)).next.bind(it); if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; return function () { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
var logger = (0, _lowcodeUtils.getLogger)({
  level: 'warn',
  bizName: 'shell-material'
});
var innerEditorSymbol = Symbol('editor');
var Material = exports.Material = /*#__PURE__*/function () {
  function Material(editor, workspaceMode) {
    var _this = this;
    if (workspaceMode === void 0) {
      workspaceMode = false;
    }
    this.workspaceMode = workspaceMode;
    this[innerEditorSymbol] = void 0;
    /**
     * 注册物料元数据管道函数
     * @param transducer
     * @param level
     * @param id
     */
    this.registerMetadataTransducer = function (transducer, level, id) {
      _this[_symbols.designerSymbol].componentActions.registerMetadataTransducer(transducer, level, id);
    };
    /**
     * 在设计器辅助层增加一个扩展 action
     * @param action
     */
    this.addBuiltinComponentAction = function (action) {
      _this[_symbols.designerSymbol].componentActions.addBuiltinComponentAction(action);
    };
    /**
     * 刷新 componentMetasMap，可触发模拟器里的 components 重新构建
     */
    this.refreshComponentMetasMap = function () {
      _this[_symbols.designerSymbol].refreshComponentMetasMap();
    };
    this[innerEditorSymbol] = editor;
  }

  /**
   * 获取组件 map 结构
   */
  var _proto = Material.prototype;
  /**
   * 设置「资产包」结构
   * @param assets
   * @returns
   */
  _proto.setAssets =
  /*#__PURE__*/
  function () {
    var _setAssets = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(assets) {
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return this[_symbols.editorSymbol].setAssets(assets);
          case 2:
            return _context.abrupt("return", _context.sent);
          case 3:
          case "end":
            return _context.stop();
        }
      }, _callee, this);
    }));
    function setAssets(_x) {
      return _setAssets.apply(this, arguments);
    }
    return setAssets;
  }()
  /**
   * 获取「资产包」结构
   * @returns
   */
  ;
  _proto.getAssets = function getAssets() {
    return this[_symbols.editorSymbol].get('assets');
  }

  /**
   * 加载增量的「资产包」结构，该增量包会与原有的合并
   * @param incrementalAssets
   * @returns
   */;
  _proto.loadIncrementalAssets = function loadIncrementalAssets(incrementalAssets) {
    return this[_symbols.designerSymbol].loadIncrementalAssets(incrementalAssets);
  };
  /**
   * 获取所有物料元数据管道函数
   * @returns
   */
  _proto.getRegisteredMetadataTransducers = function getRegisteredMetadataTransducers() {
    return this[_symbols.designerSymbol].componentActions.getRegisteredMetadataTransducers();
  }

  /**
   * 获取指定名称的物料元数据
   * @param componentName
   * @returns
   */;
  _proto.getComponentMeta = function getComponentMeta(componentName) {
    var innerMeta = this[_symbols.designerSymbol].getComponentMeta(componentName);
    return _model.ComponentMeta.create(innerMeta);
  }

  /**
   * create an instance of ComponentMeta by given metadata
   * @param metadata
   * @returns
   */;
  _proto.createComponentMeta = function createComponentMeta(metadata) {
    return _model.ComponentMeta.create(this[_symbols.designerSymbol].createComponentMeta(metadata));
  }

  /**
   * test if the given object is a ComponentMeta instance or not
   * @param obj
   * @returns
   */;
  _proto.isComponentMeta = function isComponentMeta(obj) {
    return (0, _lowcodeDesigner.isComponentMeta)(obj);
  }

  /**
   * 获取所有已注册的物料元数据
   * @returns
   */;
  _proto.getComponentMetasMap = function getComponentMetasMap() {
    var map = new Map();
    var originalMap = this[_symbols.designerSymbol].getComponentMetasMap();
    for (var _iterator = _createForOfIteratorHelperLoose(originalMap.keys()), _step; !(_step = _iterator()).done;) {
      var componentName = _step.value;
      map.set(componentName, this.getComponentMeta(componentName));
    }
    return map;
  };
  /**
   * 移除设计器辅助层的指定 action
   * @param name
   */
  _proto.removeBuiltinComponentAction = function removeBuiltinComponentAction(name) {
    this[_symbols.designerSymbol].componentActions.removeBuiltinComponentAction(name);
  }

  /**
   * 修改已有的设计器辅助层的指定 action
   * @param actionName
   * @param handle
   */;
  _proto.modifyBuiltinComponentAction = function modifyBuiltinComponentAction(actionName, handle) {
    this[_symbols.designerSymbol].componentActions.modifyBuiltinComponentAction(actionName, handle);
  }

  /**
   * 监听 assets 变化的事件
   * @param fn
   */;
  _proto.onChangeAssets = function onChangeAssets(fn) {
    var dispose = [
    // 设置 assets，经过 setAssets 赋值
    this[_symbols.editorSymbol].onChange('assets', fn),
    // 增量设置 assets，经过 loadIncrementalAssets 赋值
    this[_symbols.editorSymbol].eventBus.on('designer.incrementalAssetsReady', fn)];
    return function () {
      dispose.forEach(function (d) {
        return d && d();
      });
    };
  };
  _proto.addContextMenuOption = function addContextMenuOption(option) {
    this[_symbols.designerSymbol].contextMenuActions.addMenuAction(option);
  };
  _proto.removeContextMenuOption = function removeContextMenuOption(name) {
    this[_symbols.designerSymbol].contextMenuActions.removeMenuAction(name);
  };
  _proto.adjustContextMenuLayout = function adjustContextMenuLayout(fn) {
    this[_symbols.designerSymbol].contextMenuActions.adjustMenuLayout(fn);
  };
  (0, _createClass2["default"])(Material, [{
    key: _symbols.editorSymbol,
    get: function get() {
      if (this.workspaceMode) {
        return this[innerEditorSymbol];
      }
      var workspace = _lowcodeEditorCore.globalContext.get('workspace');
      if (workspace.isActive) {
        if (!workspace.window.editor) {
          logger.error('Material api 调用时机出现问题，请检查');
          return this[innerEditorSymbol];
        }
        return workspace.window.editor;
      }
      return this[innerEditorSymbol];
    }
  }, {
    key: _symbols.designerSymbol,
    get: function get() {
      return this[_symbols.editorSymbol].get('designer');
    }
  }, {
    key: "componentsMap",
    get: function get() {
      return this[_symbols.designerSymbol].componentsMap;
    }
  }]);
  return Material;
}();
