// @alilc/lowcode-shell/lib/api/plugins.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.Plugins = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _model = require("../model");
var _symbols = require("../symbols");
var innerPluginsSymbol = Symbol('plugin');
var Plugins = exports.Plugins = /*#__PURE__*/function () {
  function Plugins(plugins, workspaceMode) {
    if (workspaceMode === void 0) {
      workspaceMode = false;
    }
    this.workspaceMode = workspaceMode;
    this[innerPluginsSymbol] = void 0;
    this[innerPluginsSymbol] = plugins;
  }
  var _proto = Plugins.prototype;
  _proto.register = /*#__PURE__*/function () {
    var _register = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(pluginModel, options, registerOptions) {
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return this[_symbols.pluginsSymbol].register(pluginModel, options, registerOptions);
          case 2:
          case "end":
            return _context.stop();
        }
      }, _callee, this);
    }));
    function register(_x, _x2, _x3) {
      return _register.apply(this, arguments);
    }
    return register;
  }();
  _proto.init = /*#__PURE__*/function () {
    var _init = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(registerOptions) {
      return _regenerator["default"].wrap(function _callee2$(_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return this[_symbols.pluginsSymbol].init(registerOptions);
          case 2:
          case "end":
            return _context2.stop();
        }
      }, _callee2, this);
    }));
    function init(_x4) {
      return _init.apply(this, arguments);
    }
    return init;
  }();
  _proto.getPluginPreference = function getPluginPreference(pluginName) {
    return this[_symbols.pluginsSymbol].getPluginPreference(pluginName);
  };
  _proto.get = function get(pluginName) {
    var instance = this[_symbols.pluginsSymbol].get(pluginName);
    if (instance) {
      return new _model.PluginInstance(instance);
    }
    return null;
  };
  _proto.getAll = function getAll() {
    var _this$pluginsSymbol$g;
    return (_this$pluginsSymbol$g = this[_symbols.pluginsSymbol].getAll()) === null || _this$pluginsSymbol$g === void 0 ? void 0 : _this$pluginsSymbol$g.map(function (d) {
      return new _model.PluginInstance(d);
    });
  };
  _proto.has = function has(pluginName) {
    return this[_symbols.pluginsSymbol].has(pluginName);
  };
  _proto["delete"] = /*#__PURE__*/function () {
    var _delete2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(pluginName) {
      return _regenerator["default"].wrap(function _callee3$(_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return this[_symbols.pluginsSymbol]["delete"](pluginName);
          case 2:
            return _context3.abrupt("return", _context3.sent);
          case 3:
          case "end":
            return _context3.stop();
        }
      }, _callee3, this);
    }));
    function _delete(_x5) {
      return _delete2.apply(this, arguments);
    }
    return _delete;
  }();
  _proto.toProxy = function toProxy() {
    return new Proxy(this, {
      get: function get(target, prop, receiver) {
        var _target = target[_symbols.pluginsSymbol];
        if (_target.pluginsMap.has(prop)) {
          var _target$pluginsMap$ge;
          // 禁用态的插件，直接返回 undefined
          if (_target.pluginsMap.get(prop).disabled) {
            return undefined;
          }
          return (_target$pluginsMap$ge = _target.pluginsMap.get(prop)) === null || _target$pluginsMap$ge === void 0 ? void 0 : _target$pluginsMap$ge.toProxy();
        }
        return Reflect.get(target, prop, receiver);
      }
    });
  };
  (0, _createClass2["default"])(Plugins, [{
    key: _symbols.pluginsSymbol,
    get: function get() {
      if (this.workspaceMode) {
        return this[innerPluginsSymbol];
      }
      var workspace = _lowcodeEditorCore.globalContext.get('workspace');
      if (workspace.isActive) {
        return workspace.window.innerPlugins;
      }
      return this[innerPluginsSymbol];
    }
  }]);
  return Plugins;
}();
