// @alilc/lowcode-shell/lib/model/history.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.History = void 0;
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _symbols = require("../symbols");
var History = exports.History = /*#__PURE__*/function () {
  function History(document) {
    this[_symbols.documentSymbol] = void 0;
    this[_symbols.documentSymbol] = document;
  }

  /**
   * 历史记录跳转到指定位置
   * @param cursor
   */
  var _proto = History.prototype;
  _proto.go = function go(cursor) {
    this[_symbols.historySymbol].go(cursor);
  }

  /**
   * 历史记录后退
   */;
  _proto.back = function back() {
    this[_symbols.historySymbol].back();
  }

  /**
   * 历史记录前进
   */;
  _proto.forward = function forward() {
    this[_symbols.historySymbol].forward();
  }

  /**
   * 保存当前状态
   */;
  _proto.savePoint = function savePoint() {
    this[_symbols.historySymbol].savePoint();
  }

  /**
   * 当前是否是「保存点」，即是否有状态变更但未保存
   * @returns
   */;
  _proto.isSavePoint = function isSavePoint() {
    return this[_symbols.historySymbol].isSavePoint();
  }

  /**
   * 获取 state，判断当前是否为「可回退」、「可前进」的状态
   * @returns
   */;
  _proto.getState = function getState() {
    return this[_symbols.historySymbol].getState();
  }

  /**
   * 监听 state 变更事件
   * @param func
   * @returns
   */;
  _proto.onChangeState = function onChangeState(func) {
    return this[_symbols.historySymbol].onChangeState(func);
  }

  /**
   * 监听历史记录游标位置变更事件
   * @param func
   * @returns
   */;
  _proto.onChangeCursor = function onChangeCursor(func) {
    return this[_symbols.historySymbol].onChangeCursor(func);
  };
  (0, _createClass2["default"])(History, [{
    key: _symbols.historySymbol,
    get: function get() {
      return this[_symbols.documentSymbol].getHistory();
    }
  }]);
  return History;
}();
