// @alilc/lowcode-shell/lib/model/modal-nodes-manager.js
"use strict";

exports.__esModule = true;
exports.ModalNodesManager = void 0;
var _node = require("./node");
var _symbols = require("../symbols");
var ModalNodesManager = exports.ModalNodesManager = /*#__PURE__*/function () {
  function ModalNodesManager(modalNodesManager) {
    this[_symbols.modalNodesManagerSymbol] = void 0;
    this[_symbols.modalNodesManagerSymbol] = modalNodesManager;
  }
  ModalNodesManager.create = function create(modalNodesManager) {
    if (!modalNodesManager) {
      return null;
    }
    return new ModalNodesManager(modalNodesManager);
  }

  /**
   * 设置模态节点，触发内部事件
   */;
  var _proto = ModalNodesManager.prototype;
  _proto.setNodes = function setNodes() {
    this[_symbols.modalNodesManagerSymbol].setNodes();
  }

  /**
   * 获取模态节点（们）
   */;
  _proto.getModalNodes = function getModalNodes() {
    var innerNodes = this[_symbols.modalNodesManagerSymbol].getModalNodes();
    var shellNodes = [];
    innerNodes === null || innerNodes === void 0 ? void 0 : innerNodes.forEach(function (node) {
      var shellNode = _node.Node.create(node);
      if (shellNode) {
        shellNodes.push(shellNode);
      }
    });
    return shellNodes;
  }

  /**
   * 获取当前可见的模态节点
   */;
  _proto.getVisibleModalNode = function getVisibleModalNode() {
    return _node.Node.create(this[_symbols.modalNodesManagerSymbol].getVisibleModalNode());
  }

  /**
   * 隐藏模态节点（们）
   */;
  _proto.hideModalNodes = function hideModalNodes() {
    this[_symbols.modalNodesManagerSymbol].hideModalNodes();
  }

  /**
   * 设置指定节点为可见态
   * @param node Node
   */;
  _proto.setVisible = function setVisible(node) {
    this[_symbols.modalNodesManagerSymbol].setVisible(node[_symbols.nodeSymbol]);
  }

  /**
   * 设置指定节点为不可见态
   * @param node Node
   */;
  _proto.setInvisible = function setInvisible(node) {
    this[_symbols.modalNodesManagerSymbol].setInvisible(node[_symbols.nodeSymbol]);
  };
  return ModalNodesManager;
}();
