// @alilc/lowcode-shell/lib/model/setting-top-entry.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.SettingTopEntry = void 0;
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _symbols = require("../symbols");
var _node = require("./node");
var _settingField = require("./setting-field");
var SettingTopEntry = exports.SettingTopEntry = /*#__PURE__*/function () {
  function SettingTopEntry(prop) {
    this[_symbols.settingTopEntrySymbol] = void 0;
    this[_symbols.settingTopEntrySymbol] = prop;
  }
  SettingTopEntry.create = function create(prop) {
    return new SettingTopEntry(prop);
  }

  /**
   * 返回所属的节点实例
   */;
  var _proto = SettingTopEntry.prototype;
  /**
   * 获取子级属性对象
   * @param propName
   * @returns
   */
  _proto.get = function get(propName) {
    return _settingField.SettingField.create(this[_symbols.settingTopEntrySymbol].get(propName));
  }

  /**
   * @deprecated use .node instead
   */;
  _proto.getNode = function getNode() {
    return this.node;
  }

  /**
   * 获取指定 propName 的值
   * @param propName
   * @returns
   */;
  _proto.getPropValue = function getPropValue(propName) {
    return this[_symbols.settingTopEntrySymbol].getPropValue(propName);
  }

  /**
   * 设置指定 propName 的值
   * @param propName
   * @param value
   */;
  _proto.setPropValue = function setPropValue(propName, value) {
    this[_symbols.settingTopEntrySymbol].setPropValue(propName, value);
  };
  _proto.clearPropValue = function clearPropValue(propName) {
    this[_symbols.settingTopEntrySymbol].clearPropValue(propName);
  };
  (0, _createClass2["default"])(SettingTopEntry, [{
    key: "node",
    get: function get() {
      return _node.Node.create(this[_symbols.settingTopEntrySymbol].getNode());
    }
  }]);
  return SettingTopEntry;
}();
