// @alilc/lowcode-shell/lib/model/setting-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.SettingField = void 0;
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _lowcodeDesigner = require("@alilc/lowcode-designer");
var _symbols = require("../symbols");
var _node = require("./node");
var _settingTopEntry = require("./setting-top-entry");
var _componentMeta = require("./component-meta");
var _lowcodeUtils = require("@alilc/lowcode-utils");
var SettingField = exports.SettingField = /*#__PURE__*/function () {
  function SettingField(prop) {
    this[_symbols.settingFieldSymbol] = void 0;
    this[_symbols.settingFieldSymbol] = prop;
  }
  SettingField.create = function create(prop) {
    return new SettingField(prop);
  }

  /**
   * 获取设置属性的 isGroup
   */;
  var _proto = SettingField.prototype;
  /**
   * 设置 key 值
   * @param key
   */
  _proto.setKey = function setKey(key) {
    this[_symbols.settingFieldSymbol].setKey(key);
  }

  /**
   * @deprecated use .node instead
   */;
  _proto.getNode = function getNode() {
    return this.node;
  }

  /**
   * @deprecated use .parent instead
   */;
  _proto.getParent = function getParent() {
    return this.parent;
  }

  /**
   * 设置值
   * @param val 值
   */;
  _proto.setValue = function setValue(val, extraOptions) {
    this[_symbols.settingFieldSymbol].setValue(val, false, false, extraOptions);
  }

  /**
   * 设置子级属性值
   * @param propName 子属性名
   * @param value 值
   */;
  _proto.setPropValue = function setPropValue(propName, value) {
    this[_symbols.settingFieldSymbol].setPropValue(propName, value);
  }

  /**
   * 清空指定属性值
   * @param propName
   */;
  _proto.clearPropValue = function clearPropValue(propName) {
    this[_symbols.settingFieldSymbol].clearPropValue(propName);
  }

  /**
   * 获取配置的默认值
   * @returns
   */;
  _proto.getDefaultValue = function getDefaultValue() {
    return this[_symbols.settingFieldSymbol].getDefaultValue();
  }

  /**
   * 获取值
   * @returns
   */;
  _proto.getValue = function getValue() {
    return this[_symbols.settingFieldSymbol].getValue();
  }

  /**
   * 获取子级属性值
   * @param propName 子属性名
   * @returns
   */;
  _proto.getPropValue = function getPropValue(propName) {
    return this[_symbols.settingFieldSymbol].getPropValue(propName);
  }

  /**
   * 获取顶层附属属性值
   */;
  _proto.getExtraPropValue = function getExtraPropValue(propName) {
    return this[_symbols.settingFieldSymbol].getExtraPropValue(propName);
  }

  /**
   * 设置顶层附属属性值
   */;
  _proto.setExtraPropValue = function setExtraPropValue(propName, value) {
    this[_symbols.settingFieldSymbol].setExtraPropValue(propName, value);
  }

  /**
   * 获取设置属性集
   * @returns
   */;
  _proto.getProps = function getProps() {
    return _settingTopEntry.SettingTopEntry.create(this[_symbols.settingFieldSymbol].getProps());
  }

  /**
   * 是否绑定了变量
   * @returns
   */;
  _proto.isUseVariable = function isUseVariable() {
    return this[_symbols.settingFieldSymbol].isUseVariable();
  }

  /**
   * 设置绑定变量
   * @param flag
   */;
  _proto.setUseVariable = function setUseVariable(flag) {
    this[_symbols.settingFieldSymbol].setUseVariable(flag);
  }

  /**
   * 创建一个设置 field 实例
   * @param config
   * @returns
   */;
  _proto.createField = function createField(config) {
    return SettingField.create(this[_symbols.settingFieldSymbol].createField(config));
  }

  /**
   * 获取值，当为变量时，返回 mock
   * @returns
   */;
  _proto.getMockOrValue = function getMockOrValue() {
    return this[_symbols.settingFieldSymbol].getMockOrValue();
  }

  /**
   * 销毁当前 field 实例
   */;
  _proto.purge = function purge() {
    this[_symbols.settingFieldSymbol].purge();
  }

  /**
   * 移除当前 field 实例
   */;
  _proto.remove = function remove() {
    this[_symbols.settingFieldSymbol].remove();
  }

  /**
   * 设置 autorun
   * @param action
   * @returns
   */;
  _proto.onEffect = function onEffect(action) {
    return this[_symbols.settingFieldSymbol].onEffect(action);
  }

  /**
   * 返回 shell 模型，兼容某些场景下 field 已经是 shell field 了
   * @returns
   */;
  _proto.internalToShellField = function internalToShellField() {
    return this;
  };
  (0, _createClass2["default"])(SettingField, [{
    key: "isGroup",
    get: function get() {
      return this[_symbols.settingFieldSymbol].isGroup;
    }

    /**
     * 获取设置属性的 id
     */
  }, {
    key: "id",
    get: function get() {
      return this[_symbols.settingFieldSymbol].id;
    }

    /**
     * 获取设置属性的 name
     */
  }, {
    key: "name",
    get: function get() {
      return this[_symbols.settingFieldSymbol].name;
    }

    /**
     * 获取设置属性的 key
     */
  }, {
    key: "key",
    get: function get() {
      return this[_symbols.settingFieldSymbol].getKey();
    }

    /**
     * 获取设置属性的 path
     */
  }, {
    key: "path",
    get: function get() {
      return this[_symbols.settingFieldSymbol].path;
    }

    /**
     * 获取设置属性的 title
     */
  }, {
    key: "title",
    get: function get() {
      return this[_symbols.settingFieldSymbol].title;
    }

    /**
     * 获取设置属性的 setter
     */
  }, {
    key: "setter",
    get: function get() {
      return this[_symbols.settingFieldSymbol].setter;
    }

    /**
     * 获取设置属性的 expanded
     */
  }, {
    key: "expanded",
    get: function get() {
      return this[_symbols.settingFieldSymbol].expanded;
    }

    /**
     * 获取设置属性的 extraProps
     */
  }, {
    key: "extraProps",
    get: function get() {
      return this[_symbols.settingFieldSymbol].extraProps;
    }
  }, {
    key: "props",
    get: function get() {
      return _settingTopEntry.SettingTopEntry.create(this[_symbols.settingFieldSymbol].props);
    }

    /**
     * 获取设置属性对应的节点实例
     */
  }, {
    key: "node",
    get: function get() {
      return _node.Node.create(this[_symbols.settingFieldSymbol].getNode());
    }

    /**
     * 获取设置属性的父设置属性
     */
  }, {
    key: "parent",
    get: function get() {
      if ((0, _lowcodeDesigner.isSettingField)(this[_symbols.settingFieldSymbol].parent)) {
        return SettingField.create(this[_symbols.settingFieldSymbol].parent);
      }
      return _settingTopEntry.SettingTopEntry.create(this[_symbols.settingFieldSymbol].parent);
    }

    /**
     * 获取顶级设置属性
     */
  }, {
    key: "top",
    get: function get() {
      return _settingTopEntry.SettingTopEntry.create(this[_symbols.settingFieldSymbol].top);
    }

    /**
     * 是否是 SettingField 实例
     */
  }, {
    key: "isSettingField",
    get: function get() {
      return this[_symbols.settingFieldSymbol].isSettingField;
    }

    /**
     * componentMeta
     */
  }, {
    key: "componentMeta",
    get: function get() {
      return _componentMeta.ComponentMeta.create(this[_symbols.settingFieldSymbol].componentMeta);
    }

    /**
     * 获取设置属性的 items
     */
  }, {
    key: "items",
    get: function get() {
      var _this$settingFieldSym;
      return (_this$settingFieldSym = this[_symbols.settingFieldSymbol].items) === null || _this$settingFieldSym === void 0 ? void 0 : _this$settingFieldSym.map(function (item) {
        if ((0, _lowcodeUtils.isCustomView)(item)) {
          return item;
        }
        return item.internalToShellField();
      });
    }
  }]);
  return SettingField;
}();
