// @alilc/lowcode-shell/lib/model/skeleton-item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.SkeletonItem = void 0;
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _symbols = require("../symbols");
var SkeletonItem = exports.SkeletonItem = /*#__PURE__*/function () {
  function SkeletonItem(skeletonItem) {
    this[_symbols.skeletonItemSymbol] = void 0;
    this[_symbols.skeletonItemSymbol] = skeletonItem;
  }
  var _proto = SkeletonItem.prototype;
  _proto.disable = function disable() {
    var _this$skeletonItemSym, _this$skeletonItemSym2;
    (_this$skeletonItemSym = (_this$skeletonItemSym2 = this[_symbols.skeletonItemSymbol]).disable) === null || _this$skeletonItemSym === void 0 ? void 0 : _this$skeletonItemSym.call(_this$skeletonItemSym2);
  };
  _proto.enable = function enable() {
    var _this$skeletonItemSym3, _this$skeletonItemSym4;
    (_this$skeletonItemSym3 = (_this$skeletonItemSym4 = this[_symbols.skeletonItemSymbol]).enable) === null || _this$skeletonItemSym3 === void 0 ? void 0 : _this$skeletonItemSym3.call(_this$skeletonItemSym4);
  };
  _proto.hide = function hide() {
    this[_symbols.skeletonItemSymbol].hide();
  };
  _proto.show = function show() {
    this[_symbols.skeletonItemSymbol].show();
  };
  _proto.toggle = function toggle() {
    this[_symbols.skeletonItemSymbol].toggle();
  };
  (0, _createClass2["default"])(SkeletonItem, [{
    key: "name",
    get: function get() {
      return this[_symbols.skeletonItemSymbol].name;
    }
  }, {
    key: "visible",
    get: function get() {
      return this[_symbols.skeletonItemSymbol].visible;
    }
  }]);
  return SkeletonItem;
}();
