// @alilc/lowcode-editor-core/lib/utils/obx.js
"use strict";

exports.__esModule = true;
exports.untracked = exports.runInAction = exports.reaction = exports.obx = exports.observe = exports.observable = exports.mobx = exports.makeObservable = exports.makeAutoObservable = exports.computed = exports.autorun = exports.action = void 0;
var _mobxReact = require("mobx-react");
exports.observer = _mobxReact.observer;
var _mobx2 = _interopRequireWildcard(require("mobx"));
var _mobx = _mobx2;
exports.obx = _mobx2.observable;
exports.observable = _mobx2.observable;
exports.observe = _mobx2.observe;
exports.autorun = _mobx2.autorun;
exports.makeObservable = _mobx2.makeObservable;
exports.makeAutoObservable = _mobx2.makeAutoObservable;
exports.reaction = _mobx2.reaction;
exports.computed = _mobx2.computed;
exports.action = _mobx2.action;
exports.runInAction = _mobx2.runInAction;
exports.untracked = _mobx2.untracked;
exports.mobx = _mobx2;
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
(0, _mobx2.configure)({
  enforceActions: 'never'
});

// 常用的直接导出，其他的以 mobx 命名空间导出
