// @alilc/lowcode-designer/lib/designer/drag-ghost/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports["default"] = void 0;
var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));
var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));
var _react = require("react");
var _lowcodeEditorCore = require("@alilc/lowcode-editor-core");
var _dragon = require("../dragon");
var _simulator = require("../../simulator");
require("./ghost.less");
var _dec, _dec2, _dec3, _class, _class2, _descriptor, _descriptor2, _descriptor3, _descriptor4;
var DragGhost = exports["default"] = (_dec = _lowcodeEditorCore.obx.ref, _dec2 = _lowcodeEditorCore.obx.ref, _dec3 = _lowcodeEditorCore.obx.ref, (0, _lowcodeEditorCore.observer)(_class = (_class2 = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(DragGhost, _Component);
  function DragGhost(props) {
    var _this;
    _this = _Component.call(this, props) || this;
    _this.dispose = [];
    (0, _initializerDefineProperty2["default"])(_this, "titles", _descriptor, (0, _assertThisInitialized2["default"])(_this));
    (0, _initializerDefineProperty2["default"])(_this, "x", _descriptor2, (0, _assertThisInitialized2["default"])(_this));
    (0, _initializerDefineProperty2["default"])(_this, "y", _descriptor3, (0, _assertThisInitialized2["default"])(_this));
    (0, _initializerDefineProperty2["default"])(_this, "isAbsoluteLayoutContainer", _descriptor4, (0, _assertThisInitialized2["default"])(_this));
    _this.dragon = _this.props.designer.dragon;
    (0, _lowcodeEditorCore.makeObservable)((0, _assertThisInitialized2["default"])(_this));
    _this.dispose = [_this.dragon.onDragstart(function (e) {
      if (e.originalEvent.type.slice(0, 4) === 'drag') {
        return;
      }
      _this.titles = _this.getTitles(e.dragObject);
      _this.x = e.globalX;
      _this.y = e.globalY;
    }), _this.dragon.onDrag(function (e) {
      _this.x = e.globalX;
      _this.y = e.globalY;
      if ((0, _simulator.isSimulatorHost)(e.sensor)) {
        var container = e.sensor.getDropContainer(e);
        if (container !== null && container !== void 0 && container.container.componentMeta.advanced.isAbsoluteLayoutContainer) {
          _this.isAbsoluteLayoutContainer = true;
          return;
        }
      }
      _this.isAbsoluteLayoutContainer = false;
    }), _this.dragon.onDragend(function () {
      _this.titles = null;
      _this.x = 0;
      _this.y = 0;
    })];
    return _this;
  }
  var _proto = DragGhost.prototype;
  _proto.getTitles = function getTitles(dragObject) {
    var _this2 = this;
    if ((0, _dragon.isDragNodeObject)(dragObject)) {
      return dragObject.nodes.map(function (node) {
        return node.title;
      });
    }
    var dataList = Array.isArray(dragObject.data) ? dragObject.data : [dragObject.data];
    return dataList.map(function (item, i) {
      return _this2.props.designer.getComponentMeta(item.componentName).title;
    });
  };
  _proto.componentWillUnmount = function componentWillUnmount() {
    if (this.dispose) {
      this.dispose.forEach(function (off) {
        return off();
      });
    }
  };
  _proto.renderGhostGroup = function renderGhostGroup() {
    var _this$titles;
    return (_this$titles = this.titles) === null || _this$titles === void 0 ? void 0 : _this$titles.map(function (title, i) {
      var ghost = /*#__PURE__*/React.createElement("div", {
        className: "lc-ghost",
        key: i
      }, /*#__PURE__*/React.createElement(_lowcodeEditorCore.Title, {
        title: title
      }));
      return ghost;
    });
  };
  _proto.render = function render() {
    if (!this.titles || !this.titles.length) {
      return null;
    }
    if (this.isAbsoluteLayoutContainer) {
      return null;
    }
    return /*#__PURE__*/React.createElement("div", {
      className: "lc-ghost-group",
      style: {
        left: this.x,
        top: this.y
      }
    }, this.renderGhostGroup());
  };
  return DragGhost;
}(_react.Component), (_descriptor = (0, _applyDecoratedDescriptor2["default"])(_class2.prototype, "titles", [_dec], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return null;
  }
}), _descriptor2 = (0, _applyDecoratedDescriptor2["default"])(_class2.prototype, "x", [_dec2], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return 0;
  }
}), _descriptor3 = (0, _applyDecoratedDescriptor2["default"])(_class2.prototype, "y", [_dec3], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return 0;
  }
}), _descriptor4 = (0, _applyDecoratedDescriptor2["default"])(_class2.prototype, "isAbsoluteLayoutContainer", [_lowcodeEditorCore.obx], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return false;
  }
})), _class2)) || _class);
