// @alilc/lowcode-plugin-code-editor/lib/utils/babel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.availablePresets = exports.availablePlugins = void 0;
exports.registerPreset = registerPreset;
exports.registerPresets = registerPresets;
exports.transform = transform;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _core = require("@babel/core");
var _presetEnv = _interopRequireDefault(require("@babel/preset-env"));
var _presetReact = _interopRequireDefault(require("@babel/preset-react"));
var isArray = Array.isArray || function (arg) {
  return Object.prototype.toString.call(arg) === '[object Array]';
};
var availablePresets = exports.availablePresets = {};
var availablePlugins = exports.availablePlugins = {};
function loadBuiltin(builtinTable, name) {
  if (isArray(name) && typeof name[0] === 'string') {
    if (Object.prototype.hasOwnProperty.call(builtinTable, name[0])) {
      return [builtinTable[name[0]]].concat(name.slice(1));
    }
    return;
  } else if (typeof name === 'string') {
    return builtinTable[name];
  }
  // Could be an actual preset/plugin module
  return name;
}
function processOptions(options) {
  // Parse preset names
  var presets = (options.presets || []).map(function (presetName) {
    var preset = loadBuiltin(availablePresets, presetName);
    if (preset) {
      // workaround for babel issue
      // at some point, babel copies the preset, losing the non-enumerable
      // buildPreset key; convert it into an enumerable key.
      if (isArray(preset) && typeof preset[0] === 'object' && Object.prototype.hasOwnProperty.call(preset[0], 'buildPreset')) {
        preset[0] = (0, _extends2["default"])({}, preset[0], {
          buildPreset: preset[0].buildPreset
        });
      }
    } else {
      throw new Error("Invalid preset specified in Babel options: \"" + presetName + "\"");
    }
    return preset;
  });

  // Parse plugin names
  var plugins = (options.plugins || []).map(function (pluginName) {
    var plugin = loadBuiltin(availablePlugins, pluginName);
    if (!plugin) {
      throw new Error("Invalid plugin specified in Babel options: \"" + pluginName + "\"");
    }
    return plugin;
  });
  return (0, _extends2["default"])({
    babelrc: false
  }, options, {
    presets: presets,
    plugins: plugins
  });
}
function registerPreset(name, preset) {
  if (Object.prototype.hasOwnProperty.call(availablePresets, name)) {
    if (name === 'env') {
      console.warn('@babel/preset-env is now included in @babel/standalone, please remove @babel/preset-env-standalone');
    } else {
      console.warn("A preset named \"" + name + "\" is already registered, it will be overridden");
    }
  }
  availablePresets[name] = preset;
}
function transform(code, options) {
  return (0, _core.transform)(code, processOptions(options));
}
function registerPresets(newPresets) {
  Object.keys(newPresets).forEach(function (name) {
    return registerPreset(name, newPresets[name]);
  });
}
registerPresets({
  env: _presetEnv["default"],
  // es2016: () => {
  //   return {
  //     plugins: [availablePlugins["transform-exponentiation-operator"]],
  //   };
  // },
  // es2017: () => {
  //   return {
  //     plugins: [availablePlugins["transform-async-to-generator"]],
  //   };
  // },
  react: _presetReact["default"]
  // "stage-0": presetStage0,
  // "stage-1": presetStage1,
  // "stage-2": presetStage2,
  // "stage-3": presetStage3,
  // "es2015-loose": {
  //   presets: [[preset2015, { loose: true }]],
  // },
  // ES2015 preset with es2015-modules-commonjs removed
  // "es2015-no-commonjs": {
  //   presets: [[preset2015, { modules: false }]],
  // },
  // typescript: presetTypescript,
  // flow: presetFlow,
});
