// @alilc/lowcode-plugin-code-editor/lib/components/ErrorBoundary/ErrorBoundary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.ErrorBoundary = void 0;
var _message = _interopRequireDefault(require("@alifd/next/lib/message"));
var _button = _interopRequireDefault(require("@alifd/next/lib/button"));
var _loading = _interopRequireDefault(require("@alifd/next/lib/loading"));
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = _interopRequireWildcard(require("react"));
require("./ErrorBoundary.less");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
var ErrorBoundary = exports.ErrorBoundary = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(ErrorBoundary, _Component);
  function ErrorBoundary() {
    var _this;
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      hasError: false,
      info: null
    };
    _this._handleReset = function () {
      _this.setState({
        hasError: false
      });
    };
    return _this;
  }
  ErrorBoundary.getDerivedStateFromError = function getDerivedStateFromError() {
    return {
      hasError: true
    };
  };
  var _proto = ErrorBoundary.prototype;
  _proto.componentDidCatch = function componentDidCatch(error, info) {
    var onCatch = this.props.onCatch;
    if (onCatch) {
      onCatch(error, info);
      this.setState({
        info: info
      });
    }
  };
  _proto.render = function render() {
    var children = this.props.children;
    var _this$state = this.state,
      hasError = _this$state.hasError,
      info = _this$state.info;
    // const error = new Error('component render error');
    var errorInfo = {
      stack: (info === null || info === void 0 ? void 0 : info.componentStack) || ''
    };
    if (!hasError) {
      return /*#__PURE__*/_react["default"].createElement(_react.Suspense, {
        fallback: /*#__PURE__*/_react["default"].createElement(_loading["default"], {
          visible: true
        })
      }, children);
    }
    return /*#__PURE__*/_react["default"].createElement(_message["default"], {
      title: "\u51FA\u9519\u4E86~",
      type: "error"
    }, /*#__PURE__*/_react["default"].createElement("p", null, "\u8BE6\u7EC6\u9519\u8BEF: ", errorInfo || '未知原因'), /*#__PURE__*/_react["default"].createElement("div", {
      className: "plugin-code-editor-errorBoundary-actions"
    }, /*#__PURE__*/_react["default"].createElement(_button["default"], {
      onClick: this._handleReset,
      size: "small"
    }, "\u91CD\u8BD5")));
  };
  return ErrorBoundary;
}(_react.Component);
