// @alilc/lowcode-plugin-code-generator/lib/components/GravityPreview/GravitySdk.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.change = change;
exports.clear = clear;
exports["default"] = void 0;
exports.getBoostState = getBoostState;
exports.getHeight = getHeight;
exports.isPrivateMode = isPrivateMode;
exports.refresh = refresh;
exports.toggleBoostState = toggleBoostState;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// const GRAVITY_PREFIX = 'lowcodeGravityPreview';
// const GRAVITY_PREFIX = 'gravityDemoSdk';
var GRAVITY_PREFIX = 'gravityRiddleSdk';
var defaultProps = {
  // src: "https://gw.alipayobjects.com/as/g/Gravity/gravity/5.0.0-beta.10/gravityDemoSdk/index.html"
  src: 'https://gw.alipayobjects.com/as/g/Gravity/gravity/5.0.0-beta.10/gravityRiddleSdk/index.html',
  code: {},
  target: null,
  style: {
    backgroundColor: '#fff',
    border: 'none',
    position: null,
    display: 'block',
    overflow: null
  },
  scrolling: null,
  importance: null,
  sandbox: null,
  loading: null,
  styles: null,
  name: null,
  className: null,
  referrerPolicy: null,
  title: null,
  allow: 'clipboard-read;clipboard-write;camera;microphone',
  id: null,
  'aria-labelledby': null,
  'aria-hidden': null,
  'aria-label': null,
  width: null,
  height: null,
  onLoad: null,
  onMouseOver: null,
  onMouseOut: null
};
var height;
var boostState;
var iframe;
var isPrivate;

function getHeight() {
  return height;
}

function refresh() {
  var channel = GRAVITY_PREFIX;

  if (iframe) {
    iframe.current.contentWindow.postMessage({
      type: channel + "_reload"
    }, '*');
  }
}

function clear() {
  var channel = GRAVITY_PREFIX;

  if (iframe) {
    iframe.current.contentWindow.postMessage({
      type: channel + "_remove_cache"
    }, '*');
  }
}

function getBoostState() {
  return boostState;
}

function toggleBoostState() {
  var channel = GRAVITY_PREFIX;

  if (iframe) {
    iframe.current.contentWindow.postMessage({
      type: channel + "_toggle_boost_state"
    }, '*');
  }
}

function isPrivateMode() {
  return isPrivate;
}

function change(filename, content) {
  var channel = GRAVITY_PREFIX;

  if (iframe) {
    iframe.current.contentWindow.postMessage({
      type: channel + "_file_change",
      filename: filename,
      content: content
    }, '*');
  }
}

var _default = function _default(props) {
  var iframeEl = (0, _react.useRef)(null);
  iframe = iframeEl; // let iframeElOnload = useRef(false);

  (0, _react.useEffect)(function () {
    // iframeEl.current.contentWindow.location.reload();
    var channel = GRAVITY_PREFIX;
    iframeEl.current.contentWindow.postMessage({
      type: channel + "_reload"
    }, '*');

    function handler(msg) {
      var channel = GRAVITY_PREFIX;
      var type = msg.data.type;

      if (type) {
        if (type === channel + "_force_update") {
          if (props.force) {
            props.force();
          }
        }

        if (type === channel + "_code_fetch") {
          if (props.code && props.code.modules) {
            var modules = props.code.modules;
            var imn = '';
            var im = Object.keys(modules).some(function (m) {
              if (modules[m].code === null || modules[m].code === undefined) {
                imn = m;
                return m;
              }

              return false;
            });

            if (!im) {
              iframeEl.current.contentWindow.postMessage({
                type: channel + "_code_fetch",
                code: props.code
              }, '*');
            } else {
              console.warn("\u6A21\u5757 " + imn + " \u4EE3\u7801\u5B58\u5728\u5F02\u5E38\uFF0Ccode \u4E3A " + modules[imn].code);
            }
          }
        }

        if (type === channel + "_get_height") {
          height = msg.data && msg.data.height || 0;
        }

        if (type === channel + "_boost_state") {
          boostState = msg.data && msg.data.boostState;
        }

        if (type === channel + "_is_private_mode") {
          isPrivate = msg.data && msg.data.isPrivateMode;
        }
      }
    }

    window.addEventListener('message', handler);
    return function () {
      window.removeEventListener('message', handler);
    }; // return iframeEl.current.addEventListener('load', () => {
    //   iframeElOnload = true;
    //   iframeEl.current.contentWindow.postMessage({
    //     type: 'demo_fetch_code',
    //     code,
    //   });
    // });
  }, [props.code]); // useEffect(() => {
  //   if (iframeElOnload) {
  //     iframeEl.current.contentWindow.postMessage({
  //       type: 'demo_fetch_code',
  //       code,
  //     });
  //   }
  // }, [
  //   props.code
  // ]);

  var newProps = (0, _extends2["default"])({}, defaultProps, props);

  var _final = Object.create(null);

  for (var _i = 0, _Object$keys = Object.keys(newProps); _i < _Object$keys.length; _i++) {
    var prop = _Object$keys[_i];

    if (newProps[prop] != null) {
      _final[prop] = newProps[prop];
    }
  }

  for (var _i2 = 0, _Object$keys2 = Object.keys(_final.style); _i2 < _Object$keys2.length; _i2++) {
    var i = _Object$keys2[_i2];

    if (_final.style[i] == null) {
      delete _final.style[i];
    }
  }

  return (
    /*#__PURE__*/
    // eslint-disable-next-line
    _react["default"].createElement("iframe", (0, _extends2["default"])({
      ref: iframeEl
    }, _final))
  );
};

exports["default"] = _default;
